{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "07893380",
   "metadata": {},
   "source": [
    "<a id=\"0\"></a> <br>\n",
    " # Table of Contents  \n",
    "1. [Modules and Global Variables](#1)     \n",
    "1. [Set up](#2)\n",
    "    1. [Variant Definitions](#3)\n",
    "1. [Data Preprocessing](#4)\n",
    "1. [Processing](#5)\n",
    "1. [Back Projection Processing](#6)\n",
    "    1. [Step 1](#7)\n",
    "    1. [Step 2](#8)\n",
    "    1. [Step 3](#9)\n",
    "1. [Trimming and Filtering](#10)\n",
    "    1. [Back-projection Trimming](#11)\n",
    "    1. [Naive Trimming](#12)\n",
    "    1. [Filtering](#13)\n",
    "        1. [Step 1, Naive](#14)\n",
    "        1. [Step 2, Filter with Masks](#15)\n",
    "        1. [Step 3, Combine Masks](#16)\n",
    "        1. [Step 4, Filter Back-Projected](#17)        \n",
    "1. [Covariance and Inference](#18)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29cc3841",
   "metadata": {},
   "source": [
    "<a id=\"1\"></a> \n",
    "## Modules and Global Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fbd12442",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python version 3.9.7 (default, Sep 16 2021, 08:50:36) \n",
      "[Clang 10.0.0 ]\n",
      "numpy version 1.20.3\n",
      "pandas version 1.3.4\n",
      "matplotlib version 3.4.3\n",
      "scipy version 1.7.1\n"
     ]
    }
   ],
   "source": [
    "# import modules and global variables\n",
    "\n",
    "import os, sys\n",
    "from imp import reload\n",
    "import datetime as dt\n",
    "import shutil\n",
    "import posixpath as ppath\n",
    "from copy import deepcopy\n",
    "import time\n",
    "import data_processing as dp\n",
    "from tqdm import tqdm\n",
    "from collections import defaultdict\n",
    "\n",
    "print('python version %s' % sys.version)\n",
    "\n",
    "import numpy as np\n",
    "print('numpy version %s' % np.__version__)\n",
    "\n",
    "import pandas as pd\n",
    "print('pandas version %s' % pd.__version__)\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "print('matplotlib version %s' % matplotlib.__version__)\n",
    "\n",
    "import scipy\n",
    "from scipy import linalg\n",
    "print('scipy version %s' % scipy.__version__)\n",
    "\n",
    "# GLOBAL VARIABLES\n",
    "\n",
    "# sequence related variables\n",
    "\n",
    "NUC = ['-', 'A', 'C', 'G', 'T']\n",
    "\n",
    "# local directories\n",
    "\n",
    "USER_NAME         = 'YOUR_LOCAL_PATH' \n",
    "\n",
    "SARS_DIR          = os.path.join(USER_NAME, 'paper-epi-backprojection/src')\n",
    "DATA_DIR          = os.path.join(SARS_DIR, 'data-processing-pipeline')\n",
    "SCRIPT_DIR        = os.path.join(DATA_DIR, 'scripts')\n",
    "INF_DIR           = os.path.join(DATA_DIR, 'inference-c++')\n",
    "BP_DIR            = os.path.join(DATA_DIR, 'deconvolution')\n",
    "\n",
    "# cluster directories\n",
    "\n",
    "DATA_DATE         = '2024-01-26'\n",
    "SSH_USER          = 'YOUR_CLUSTER_NAME'\n",
    "USER_EMAIL        = 'YOUR_EMAIL'\n",
    "SSH_HOME          = 'YOUR_SSH_SERVER'\n",
    "SSH_DATA          = ppath.join('YOUR_CLUSTER_DIRECTORY', SSH_USER)\n",
    "SSH_LMDB          = ppath.join(SSH_DATA, f'lmdb-{DATA_DATE}')\n",
    "\n",
    "# metadata variables\n",
    "\n",
    "METADATA_COLS     = [   'accession', 'virus_name',            'date', 'location',             'location_additional', 'submission_date']\n",
    "METADATA_OCOLS    = ['Accession ID', 'Virus name', 'Collection date', 'Location', 'Additional location information', 'Submission date']\n",
    "METADATA_XFORM    = [           str,    str.lower,               str,  str.lower,                         str.lower,               str]\n",
    "METADATA_COMP     = os.path.join(SARS_DIR, f'metadata_tsv_2024_01_26.tar.xz')\n",
    "METADATA_FILE     = os.path.join(SARS_DIR, f'metadata-{DATA_DATE}', 'metadata.tsv')\n",
    "METADATA_DIR      = os.path.join(SARS_DIR, f'metadata-{DATA_DATE}')\n",
    "MSA_NAME          = 'msaCodon_0126.tar.xz'\n",
    "\n",
    "METADATA_INTERIM  = 'outputmeta.csv'\n",
    "METADATA_INT_ALT  = 'merged-interim.csv'\n",
    "MERGED_FINAL      = f'metadata-{DATA_DATE}.csv'\n",
    "METADATA_NEW      = ppath.join(SSH_DATA, MERGED_FINAL)\n",
    "REF_TAG           = 'EPI_ISL_402125'\n",
    "REF_FILE          = f'ref-index-{DATA_DATE}'\n",
    "\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67680160",
   "metadata": {},
   "source": [
    "<a id=\"2\"></a> \n",
    "## Set up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c7354ba3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scp /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/processing-scripts/data_processing.py efinn@cluster.csb.pitt.edu: &&\n",
      "scp -r /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/processing-scripts/* efinn@cluster.csb.pitt.edu:\n"
     ]
    }
   ],
   "source": [
    "# Transfer files to the cluster\n",
    "\n",
    "print('scp %s/%s %s &&' % (SCRIPT_DIR, 'data_processing.py', SSH_HOME))\n",
    "\n",
    "file_paths = os.path.join(SCRIPT_DIR, '*')\n",
    "print(f'scp -r {file_paths} {SSH_HOME}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "381747ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir -p /net/dali/home/barton/efinn/deconvolution-new &&\n",
      "mkdir -p ./MPL/out &&\n",
      "mkdir /net/dali/home/barton/efinn/deconvolution-new/regions-times &&\n",
      "mkdir /net/dali/home/barton/efinn/deconvolution-new/msa-reg &&\n",
      "mkdir /net/dali/home/barton/efinn/deconvolution-new/genome-data &&\n",
      "mkdir /net/dali/home/barton/efinn/deconvolution-new/genome-trimmed &&\n",
      "mkdir /net/dali/home/barton/efinn/deconvolution-new/genome-filtered &&\n",
      "mkdir /net/dali/home/barton/efinn/deconvolution-new/max-freqs &&\n",
      "mkdir /net/dali/home/barton/efinn/deconvolution-new/freqs &&\n",
      "mkdir /net/dali/home/barton/efinn/deconvolution-new/counts &&\n",
      "mkdir /net/dali/home/barton/efinn/deconvolution-new/trajectories &&\n",
      "mkdir /net/dali/home/barton/efinn/deconvolution-new/infer\n"
     ]
    }
   ],
   "source": [
    "# Make files on directory\n",
    "\n",
    "# updated: list of folders used in processing\n",
    "\n",
    "print(f'mkdir -p {SSH_DATA} &&')\n",
    "#print('mkdir -p ./MPL/out &&')\n",
    "dir_list = [\n",
    "    'regions-times',\n",
    "    'msa-reg', 'mutant-counts', 'impute-gaps', 'genome-data', 'genome-unique',\n",
    "    'genome-trimmed', 'genome-filtered-05', 'genome-max-freqs', \n",
    "    'genome-max-mask-05', 'genome-max-combine-mask-05', \n",
    "    'genome-filtered-compare-05', 'genome-covar-temp', 'genome-covar', \n",
    "    'bp-covar', 'bp-filtered-compare-05', 'bp-format-sh-dir', 'bp-formatted', 'bp-submit-dir', \n",
    "    'bp-trim-sh-dir', 'bp-filter-20-05', 'logs', 'scratch', 'scratch-bp', 'bp-trim-20'\n",
    "]\n",
    "for name in dir_list:\n",
    "    dir_path = ppath.join(SSH_DATA, name)\n",
    "    if name!=dir_list[-1]:\n",
    "        print(f'mkdir {dir_path} &&')\n",
    "    else:\n",
    "        print(f'mkdir {dir_path}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df219f13",
   "metadata": {},
   "source": [
    "<a id=\"3\"></a> \n",
    "### Variant definitions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "421724d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Variant definitions\n",
    "\n",
    "epsilon        = np.unique(['NSP13-260-0-T', 'S-13-1-T', 'S-152-2-T', 'NC-28271-T', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G', \n",
    "                  'S-452-1-G', 'N-205-1-T', 'ORF3a-57-2-T', 'NSP9-65-0-G']) # Done\n",
    "\n",
    "alpha          = np.unique(['NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-107-1--', 'NSP6-107-2--', 'NSP6-108-0--',\n",
    "                  'NSP6-108-1--', 'NSP6-108-2--', 'S-501-0-T', 'NSP12-412-2-T', 'NSP2-36-2-T', 'NSP3-183-1-T', 'NSP3-890-1-A', \n",
    "                  'NSP3-1089-2-T', 'NSP3-1412-1-C', 'NSP12-613-2-T', 'NSP12-912-2-C', 'S-68-1--', 'S-68-2--', 'S-69-0--', 'S-69-1--', \n",
    "                  'S-69-2--', 'S-70-0--', 'S-143-2--', 'S-144-0--', 'S-144-1--', 'S-570-1-A', 'S-681-1-A', 'S-716-1-T', 'S-982-0-G', \n",
    "                  'S-1118-0-C', 'ORF8-27-0-T', 'ORF8-52-1-T', 'ORF8-73-1-G', 'N-3-0-C', 'N-3-1-T', 'N-3-2-A', 'N-235-1-T',\n",
    "                  'N-203-1-A', 'N-203-2-A', 'N-204-0-C', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G'])    # Done\n",
    "\n",
    "# NEED TO ADD S:D138Y\n",
    "gamma          = np.unique(['NSP9-31-2-T', 'NSP1-156-2-C', 'NSP3-10-2-T', 'NSP3-370-1-T', 'NSP3-977-0-C', 'NSP3-1200-2-G', 'NSP3-1298-2-G', \n",
    "                  'NSP12-140-2-T', 'NSP13-341-2-T', 'S-20-1-A', 'S-417-1-C', 'S-1027-1-T', 'ORF3a-253-0-C', 'ORF8-92-0-A', 'N-80-1-G', \n",
    "                  'S-190-2-T', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G', 'S-501-0-T', 'NSP6-106-0--', 'NSP6-106-1--', \n",
    "                  'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-107-1--', 'NSP6-107-2--', 'NSP6-108-0--', 'NSP6-108-1--', 'NSP6-108-2--',\n",
    "                  'N-203-1-A', 'N-203-2-A', 'N-204-0-C', 'N-202-0-T', 'N-202-1-C', 'S-18-0-T', 'S-26-0-T', 'S-655-0-T',\n",
    "                  'S-1176-0-T'])    # Done\n",
    "\n",
    "twentyE_EU1    = np.unique(['NSP16-199-2-C', 'NSP1-60-2-C', 'NSP3-1189-2-T', 'M-93-2-G', 'N-220-1-T', 'ORF10-30-0-T', 'S-222-1-T',\n",
    "                  'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G'])    # Done\n",
    "# This is Pango lineage B.1.177\n",
    "\n",
    "delta          = np.unique(['NSP12-671-0-A', 'NC-209-T', 'NSP13-77-1-T', 'S-19-1-G', 'S-156-1--', 'S-156-2--', 'S-157-0--', 'S-157-1--', 'S-157-2--', \n",
    "                  'S-158-0--', 'S-478-1-A', 'S-681-1-G', 'S-950-0-A', 'ORF3a-26-1-T', 'M-82-1-C', 'ORF7a-82-1-C', 'ORF7a-120-1-T', \n",
    "                  'ORF8-119-0--', 'ORF8-119-1--', 'ORF8-119-2--', 'ORF8-120-0--', 'ORF8-120-1--', 'ORF8-120-2--', 'N-63-1-G', 'N-203-1-T', \n",
    "                  'N-377-0-T', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G', 'S-452-1-G', 'NC-28270--', 'NSP4-492-1-T'])    # Done\n",
    "\n",
    "beta           = np.unique(['S-80-1-C', 'NSP3-837-2-T', 'S-215-1-G', 'E-71-1-T', 'S-240-2--', 'S-241-0--', 'S-242-1--', 'S-242-2--', 'S-243-0--', \n",
    "                  'S-240-1--', 'S-241-1--', 'S-241-2--', 'S-242-0--', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G', 'S-501-0-T',\n",
    "                  'NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-107-1--', 'NSP6-107-2--', 'NSP6-108-0--', \n",
    "                  'NSP6-108-1--', 'NSP6-108-2--', 'S-417-2-T', 'S-484-0-A', 'S-701-1-T', 'ORF3a-57-2-T', 'NSP2-85-1-T', 'NSP5-90-1-G',\n",
    "                  'N-205-1-T', 'NC-173-T', 'ORF8-120-2-T'])    # Done\n",
    "\n",
    "lambda_new     = np.unique(['S-246-2--', 'NSP3-1569-0-G', 'S-247-0--', 'S-247-1--', 'S-247-2--', 'S-248-0--', 'S-248-1--', 'S-248-2--', 'S-249-0--', \n",
    "                  'S-249-1--', 'S-249-2--', 'S-250-0--', 'S-250-1--', 'S-250-2--', 'S-251-0--', 'S-251-1--', 'S-251-2--', 'S-252-0--', \n",
    "                  'S-246-1--', 'S-252-1--', 'S-252-2--', 'S-253-0--', 'N-214-0-T', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G', \n",
    "                  'NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-107-1--', 'NSP6-107-2--', 'NSP6-108-0--', \n",
    "                  'NSP6-108-1--', 'NSP6-108-2--', 'S-75-1-T', 'S-76-1-T', 'S-452-1-A', 'S-490-1-C', 'S-859-1-A', 'NSP3-428-1-T',\n",
    "                  'NSP3-1469-0-T', 'NSP4-438-1-C', 'NSP4-492-1-T', 'NSP5-15-0-A']) # Done\n",
    "\n",
    "iota           = np.unique(['N-234-2-A', 'S-5-0-T', 'S-95-1-T', 'S-253-1-G', 'S-484-0-A', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G',\n",
    "                 'S-701-1-T', 'NSP13-88-2-C', 'ORF3a-42-1-T', 'ORF3a-67-2-T', 'NSP2-85-1-T', 'NSP4-438-1-C', 'NSP6-106-0--', 'NSP6-106-1--', \n",
    "                 'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-107-1--', 'NSP6-107-2--', 'NSP6-108-0--',  'NSP6-108-1--', 'NSP6-108-2--', 'N-199-1-T',\n",
    "                 'N-232-2-A', 'ORF8-11-1-T', 'NSP15-214-2-G', 'NC-28270--'])\n",
    "\n",
    "D614G          = ['NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G']    # Done\n",
    "\n",
    "B1_1_318       = np.unique(['NSP15-320-0-A', 'S-575-2-C', 'S-1238-2-A', 'ORF7b-44-2--', 'NC-27887--', 'NC-27888--', 'NC-27889--', 'NC-27890--', \n",
    "                  'NC-27891--', 'NC-27892--', 'ORF8-1-0--', 'ORF8-1-1--', 'ORF8-1-2--', 'NSP4-173-1-T', 'S-796-0-C', 'ORF8-2-0--', \n",
    "                  'ORF8-2-1--', 'ORF8-2-2--', 'ORF8-3-0--', 'ORF8-3-1--', 'NC-28270-G', 'N-208-2--', 'N-209-0--', 'N-208-1--',\n",
    "                  'NSP3-378-1-T', 'NSP3-1693-2-T', 'NSP5-21-1-T', 'NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', \n",
    "                  'NSP6-107-1--', 'NSP6-107-2--', 'NSP6-108-0--', 'NSP6-108-1--', 'NSP6-108-2--', 'S-95-1-T', 'S-143-2--', 'S-144-0--', \n",
    "                  'S-144-1--', 'S-484-0-A', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G', 'S-681-1-A', 'M-82-1-C',\n",
    "                  'N-203-1-A', 'N-203-2-A', 'N-204-0-C']) # Add N-234-2-T, Many others\n",
    "\n",
    "# need to add orf1b-314, 3037\n",
    "omicron        = np.unique(['NC-240-T', 'NSP3-38-1-G', 'NSP3-106-2-T', 'NSP3-889-2-G', 'NSP3-1265-1--', 'NSP3-1265-2--', 'NSP3-1266-0--', \n",
    "                  'NSP3-1892-0-A', 'NSP4-492-1-T', 'NSP5-132-1-A', 'NSP6-104-1--', 'NSP6-104-2--', 'NSP6-105-0--', 'NSP6-105-1--', \n",
    "                  'NSP6-105-2--', 'NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-189-0-G', 'NSP10-57-2-C', \n",
    "                  'NSP12-323-1-T', 'NSP12-600-2-T', 'S-67-1-T', 'S-68-1--', 'S-68-2--', 'S-69-0--', 'S-69-1--', \n",
    "                  'S-69-2--', 'S-70-0--', 'S-95-1-T', 'S-142-1--', 'S-142-2--', 'S-143-0--', 'S-143-1--', 'S-143-2--', 'S-144-0--', \n",
    "                  'S-144-1--', 'S-144-2--', 'S-145-0--', 'S-211-1--', 'S-211-2--', 'S-212-0--', 'S-339-1-A', 'S-371-0-C', \n",
    "                  'S-371-1-T', 'S-373-0-C', 'S-417-2-T', 'S-440-2-G', 'S-446-0-A', 'S-547-1-A', 'S-614-1-G', 'S-655-0-T', \n",
    "                  'S-679-2-G', 'S-681-1-A', 'S-764-2-A', 'S-796-0-T', 'S-856-2-A', 'S-954-2-T', 'S-969-2-A', 'S-981-0-T', \n",
    "                  'S-1146-2-T', 'ORF3a-64-2-T', 'E-9-1-T', 'M-3-1-G', 'M-19-0-G', 'M-63-0-A', 'ORF6-20-0-C', \n",
    "                  'NC-28270-T', 'N-13-1-T', 'N-30-1--', 'N-30-2--', 'N-31-0--', 'N-31-1--', 'N-31-2--', 'N-32-0--', 'N-32-1--', \n",
    "                  'N-32-2--', 'N-33-0--', 'N-203-1-A', 'N-203-2-A', 'N-204-0-C', 'S-375-1-T', 'S-477-1-A', 'S-478-1-A', \n",
    "                  'S-484-1-C', 'S-493-1-G', 'S-496-0-A', 'S-498-1-G', 'S-501-0-T', 'S-505-0-C', 'NSP14-42-0-G',\n",
    "                  'ORF7b-18-0-T', 'S-214m-2-G', 'S-214n-2-A', 'S-214o-2-G', 'S-214p-2-C', 'S-214q-2-A', 'S-214r-2-A', \n",
    "                  'S-214s-2-G', 'S-214t-2-A', 'S-214u-2-A'])  \n",
    "\n",
    "# 3037\n",
    "ba2            = np.unique(['S-339-1-A', 'S-371-1-T', 'S-373-0-C', 'S-417-2-T', 'S-440-2-G', 'S-655-0-T', \n",
    "                  'S-679-2-G', 'S-681-1-A', 'S-764-2-A', 'S-796-0-T', 'S-954-2-T', 'S-969-2-A',\n",
    "                  'N-13-1-T', 'N-30-1--', 'N-30-2--', 'N-31-0--', 'N-31-1--', 'N-31-2--', 'N-32-0--', 'N-32-1--', \n",
    "                  'N-32-2--', 'N-33-0--', 'N-203-1-A', 'N-203-2-A', 'N-204-0-C', 'NSP4-492-1-T', 'NSP5-132-1-A', \n",
    "                  'NSP6-104-1--', 'NSP6-104-2--', 'NSP6-105-0--', 'NSP6-105-1--', 'NSP6-105-2--', 'NSP6-106-0--', \n",
    "                  'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'E-9-1-T', 'ORF3a-223-1-T',\n",
    "                  'M-19-0-G', 'M-63-0-A', 'S-19-1-T', 'S-24-1--', 'S-24-2--', 'S-25-0--', 'S-25-1--', 'S-25-2--',\n",
    "                  'S-26-0--', 'S-26-1--', 'S-26-2--', 'S-27-0--', 'S-142-1-A', 'S-213-1-G', \n",
    "                  'S-375-1-T', 'S-376-0-G', 'S-408-2-C', 'S-477-1-A', 'S-478-1-A', 'S-484-1-C', 'S-493-1-G', \n",
    "                  'S-498-1-G', 'S-501-0-T', 'S-505-0-C', 'N-413-0-C', 'NSP13-392-0-T', 'NSP14-42-0-G',\n",
    "                  'NSP15-112-1-T', 'ORF6-61-0-C', 'ORF6-61-1-T', 'ORF6-61-2-C', 'S-1146-2-T',\n",
    "                  'NSP3-534-2-T', 'NSP4-290-2-G', 'NSP5-48-2-T', 'NSP5-131-2-A', 'NSP9-65-2-T', 'NSP12-758-2-T',\n",
    "                  'NSP15-145-2-G', 'ORF3a-64-2-T', 'M-112-2-T', 'ORF6-20-0-C', 'ORF7b-18-0-T', 'NC-28270-T', \n",
    "                  'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G', 'S-405-0-A', 'S-214m-2-G', \n",
    "                  'S-214n-2-A', 'S-214o-2-G', 'S-214p-2-C', 'S-214q-2-A', 'S-214r-2-A', 'S-214s-2-G', \n",
    "                  'S-214t-2-A', 'S-214u-2-A', 'NSP1-135-2-G', 'NSP3-24-1-T', 'NSP3-489-0-A', 'NSP4-264-0-T', \n",
    "                  'NSP4-327-1-T'])\n",
    "\n",
    "# deletion after 29734\n",
    "ba4            = np.unique(['S-339-1-A', 'S-371-1-T', 'S-373-0-C', 'S-417-2-T', 'S-440-2-G', 'S-655-0-T', \n",
    "                  'S-679-2-G', 'S-681-1-A', 'S-764-2-A', 'S-796-0-T', 'S-954-2-T', 'S-969-2-A',\n",
    "                  'S-19-1-T', 'S-24-1--', 'S-24-2--', 'S-25-0--', 'S-25-1--', 'S-25-2--', 'S-452-1-G',\n",
    "                  'S-26-0--', 'S-26-1--', 'S-26-2--', 'S-27-0--', 'S-142-1-A', 'S-213-1-G', \n",
    "                  'S-375-1-T', 'S-376-0-G', 'S-408-2-C', 'S-477-1-A', 'S-478-1-A', 'S-484-1-C',\n",
    "                  'S-498-1-G', 'S-501-0-T', 'S-505-0-C', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', \n",
    "                  'S-614-1-G', 'S-405-0-A', 'S-214m-2-G', 'S-214n-2-A', 'S-214o-2-G', \n",
    "                  'S-214p-2-C', 'S-214q-2-A', 'S-214r-2-A', 'S-214s-2-G', 'S-214t-2-A', 'S-214u-2-A',\n",
    "                  'S-68-1--', 'S-68-2--', 'S-69-0--', 'S-69-1--', 'S-69-2--', 'S-70-0--', 'NSP1-135-2-G',\n",
    "                  'N-13-1-T', 'N-30-1--', 'N-30-2--', 'N-31-0--', 'N-31-1--', 'N-31-2--', 'N-32-0--', \n",
    "                  'N-32-1--', 'N-32-2--', 'N-33-0--', 'N-203-1-A', 'N-203-2-A', 'N-204-0-C',\n",
    "                  'NSP3-24-1-T', 'NSP3-489-0-A', 'NSP4-264-0-T', 'NSP4-492-1-T', 'NSP4-327-1-T',\n",
    "                  'NSP6-104-1--', 'NSP6-104-2--', 'NSP6-105-0--', 'NSP6-105-1--', 'NSP6-105-2--', 'NSP6-106-0--', \n",
    "                  'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP5-132-1-A', 'NSP13-392-0-T', 'NSP14-42-0-G',\n",
    "                  'NSP15-145-2-G', 'ORF3a-223-1-T', 'M-19-0-G', 'M-63-0-A', 'E-9-1-T',\n",
    "                  'NSP3-534-2-T', 'NSP4-290-2-G', 'NSP5-48-2-T', 'NSP5-131-2-A', 'NSP9-65-2-T', 'NSP12-758-2-T',\n",
    "                  'NSP15-145-2-G', 'ORF3a-64-2-T', 'M-112-2-T', 'ORF6-20-0-C', 'ORF7b-18-0-T', 'NC-28270-T',\n",
    "                  'S-1146-2-T', 'ORF7b-11-2-T', 'S-486-0-G', 'NSP8-23-2-A', 'N-151-0-T',\n",
    "                  'NSP1-141-0--', 'NSP1-141-1--', 'NSP1-141-2--', 'NSP1-142-0--', 'NSP1-142-1--',\n",
    "                  'NSP1-142-2--', 'NSP1-143-0--', 'NSP1-143-1--', 'NSP1-143-2--', 'N-413-0-C',\n",
    "                  'ORF6-61-0-C', 'ORF6-61-1-T', 'ORF6-61-2-C'])\n",
    "\n",
    "# deletion after 29734\n",
    "ba5            = np.unique(['S-339-1-A', 'S-371-1-T', 'S-373-0-C', 'S-417-2-T', 'S-440-2-G', 'S-655-0-T', \n",
    "                  'S-679-2-G', 'S-681-1-A', 'S-764-2-A', 'S-796-0-T', 'S-954-2-T', 'S-969-2-A',\n",
    "                  'S-19-1-T', 'S-24-1--', 'S-24-2--', 'S-25-0--', 'S-25-1--', 'S-25-2--', 'S-452-1-G',\n",
    "                  'S-26-0--', 'S-26-1--', 'S-26-2--', 'S-27-0--', 'S-142-1-A', 'S-213-1-G', \n",
    "                  'S-375-1-T', 'S-376-0-G', 'S-408-2-C', 'S-477-1-A', 'S-478-1-A', 'S-484-1-C',\n",
    "                  'S-498-1-G', 'S-501-0-T', 'S-505-0-C', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', \n",
    "                  'S-614-1-G', 'S-405-0-A','S-214m-2-G', 'S-214n-2-A', 'S-214o-2-G', 'S-214p-2-C', \n",
    "                  'S-214q-2-A', 'S-214r-2-A', 'S-214s-2-G', 'S-214t-2-A', 'S-214u-2-A',\n",
    "                  'S-68-1--', 'S-68-2--', 'S-69-0--', 'S-69-1--', 'S-69-2--', 'S-70-0--', 'NSP1-135-2-G',\n",
    "                  'N-13-1-T', 'N-30-1--', 'N-30-2--', 'N-31-0--', 'N-31-1--', 'N-31-2--', 'N-32-0--', \n",
    "                  'N-32-1--', 'N-32-2--', 'N-33-0--', 'N-203-1-A', 'N-203-2-A', 'N-204-0-C',\n",
    "                  'NSP3-24-1-T', 'NSP3-489-0-A', 'NSP4-264-0-T', 'NSP4-492-1-T', 'NSP4-327-1-T',\n",
    "                  'NSP6-104-1--', 'NSP6-104-2--', 'NSP6-105-0--', 'NSP6-105-1--', 'NSP6-105-2--', 'NSP6-106-0--', \n",
    "                  'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP5-132-1-A', 'NSP13-392-0-T', 'NSP14-42-0-G',\n",
    "                  'NSP3-534-2-T', 'NSP4-290-2-G', 'NSP5-48-2-T', 'NSP5-131-2-A', 'NSP9-65-2-T', 'NSP12-758-2-T',\n",
    "                  'NSP15-145-2-G', 'ORF3a-223-1-T', 'M-19-0-G', 'M-63-0-A', 'E-9-1-T', 'M-3-0-A', 'S-486-0-G',\n",
    "                  'NSP8-23-2-A', 'N-413-0-C', 'S-1146-2-T', 'NC-28270-T', 'ORF3a-64-2-T', 'ORF7b-18-0-T'])\n",
    "\n",
    "ba2121         = np.unique(['S-339-1-A', 'S-371-1-T', 'S-373-0-C', 'S-417-2-T', 'S-440-2-G', 'S-655-0-T', \n",
    "                  'S-679-2-G', 'S-681-1-A', 'S-764-2-A', 'S-796-0-T', 'S-954-2-T', 'S-969-2-A',\n",
    "                  'S-19-1-T', 'S-24-1--', 'S-24-2--', 'S-25-0--', 'S-25-1--', 'S-25-2--', 'S-452-1-G',\n",
    "                  'S-26-0--', 'S-26-1--', 'S-26-2--', 'S-27-0--', 'S-142-1-A', 'S-213-1-G', \n",
    "                  'S-375-1-T', 'S-376-0-G', 'S-408-2-C', 'S-477-1-A', 'S-478-1-A', 'S-484-1-C',\n",
    "                  'S-498-1-G', 'S-501-0-T', 'S-505-0-C', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', \n",
    "                  'S-614-1-G', 'S-405-0-A', 'S-214m-2-G', 'S-214n-2-A', 'S-214o-2-G', 'S-214p-2-C', \n",
    "                  'S-214q-2-A', 'S-214r-2-A', 'S-214s-2-G', 'S-214t-2-A', 'S-214u-2-A',\n",
    "                  'S-68-1--', 'S-68-2--', 'S-69-0--', 'S-69-1--', 'S-69-2--', 'S-70-0--', 'S-452-1-A',\n",
    "                  'N-13-1-T', 'N-30-1--', 'N-30-2--', 'N-31-0--', 'N-31-1--', 'N-31-2--', 'N-32-0--', 'N-32-1--', \n",
    "                  'N-32-2--', 'N-33-0--', 'N-203-1-A', 'N-203-2-A', 'N-204-0-C', 'NSP4-492-1-T', 'NSP5-132-1-A', \n",
    "                  'NSP6-104-1--', 'NSP6-104-2--', 'NSP6-105-0--', 'NSP6-105-1--', 'NSP6-105-2--', 'NSP6-106-0--', \n",
    "                  'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', 'NSP6-189-0-G', 'E-9-1-T', 'ORF3a-223-1-T',\n",
    "                  'M-19-0-G', 'M-63-0-A', 'N-413-0-C', 'NSP13-392-0-T', 'NSP14-42-0-G',\n",
    "                  'NSP15-112-1-T', 'ORF6-61-0-C', 'ORF6-61-1-T', 'ORF6-61-2-C', 'S-1146-2-T',\n",
    "                  'NSP3-534-2-T', 'NSP4-290-2-G', 'NSP5-48-2-T', 'NSP5-131-2-A', 'NSP9-65-2-T', 'NSP12-758-2-T',\n",
    "                  'NSP15-145-2-G', 'ORF3a-64-2-T', 'M-112-2-T', 'ORF6-20-0-C', 'ORF7b-18-0-T', 'NC-28270-T',\n",
    "                  'NSP1-135-2-G', 'NSP3-24-1-T', 'NSP3-489-0-A', 'NSP4-264-0-T', 'NSP4-327-1-T', \n",
    "                  'S-704-1-T', 'S-452-1-A'])\n",
    " \n",
    "# synonymous: 9593T\n",
    "eta            = np.unique(['NSP2-231-2-T', 'NSP2-334-2-G', 'NSP2-618-2-A', 'NSP4-13-2-C', 'NSP16-22-2-G',\n",
    "                  'S-888-0-C', 'ORF6-2-0--', 'ORF6-2-1--', 'ORF6-2-2--', 'N-2-1--', 'N-2-2--',\n",
    "                  'N-3-0--', 'N-12-1-G', 'N-142-2-G', 'S-68-1--', 'S-68-2--', 'S-69-0--', 'S-69-1--', \n",
    "                  'S-69-2--', 'S-70-0--', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G',\n",
    "                  'M-82-1-C', 'NSP6-106-0--', 'NSP6-106-1--', 'NSP6-106-2--', 'NSP6-107-0--', \n",
    "                  'NSP6-107-1--', 'NSP6-107-2--', 'NSP6-108-0--',  'NSP6-108-1--', 'NSP6-108-2--',\n",
    "                  'S-52-1-G', 'S-67-1-T', 'S-143-2--', 'S-144-0--', 'S-144-1--', 'S-484-0-A', \n",
    "                  'S-677-2-C', 'N-205-1-T', 'E-21-0-T', 'NSP3-1189-1-T',\n",
    "                  'NSP14-44-2-T', 'S-1062-2-C', 'N-142-2-G', 'NC-29542-T'])\n",
    "\n",
    "# confused about orf1a-T1567I\n",
    "kappa           = np.unique(['NSP12-323-1-T', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G',\n",
    "                   'S-154-0-A', 'S-1071-2-T', 'S-681-1-G', 'S-484-0-C', 'S-452-1-G', 'NSP13-206-0-T',\n",
    "                   'NSP13-429-2-T', 'NSP15-259-1-G', 'NSP15-261-0-G', 'N-203-1-T', 'N-377-0-T',\n",
    "                   'M-82-1-G', 'ORF3a-26-1-T', 'NSP3-749-1-T', 'NSP6-77-0-G', 'ORF7a-82-1-C',\n",
    "                   'NC-209-T', 'NSP3-246-2-T', 'M-53-2-T', 'NC-28270--'])\n",
    "\n",
    "# maybe S-144-0-A, S-144-1-C, S-144-2-C, S-145-0-A (actually there is an insertion here)\n",
    "# not positive about 'NSP3-237-0-G'\n",
    "mu              = np.unique(['S-346-2-C', 'NC-240-T', 'NSP3-106-2-T', 'NSP12-323-1-T', 'S-614-1-G', 'S-95-1-T',\n",
    "                   'S-484-1-C', 'S-501-0-T', 'S-681-1-A', 'S-950-0-A', 'NSP3-237-0-G', 'NSP3-720-1-T',\n",
    "                   'NSP4-492-1-T', 'NSP6-160-1-G', 'NSP12-323-1-T', 'NSP13-419-0-T', 'N-205-1-T',\n",
    "                   'ORF3a-57-2-T', 'ORF3a-256-0--', 'ORF3a-256-1--', 'ORF3a-256-2--', 'ORF3a-257-0--',\n",
    "                   'ORF3a-257-2-C', 'ORF8-11-1-T', 'ORF8-38-0-T', 'ORF8-67-1-T',\n",
    "                   'NSP3-1106-2-T', 'NSP10-11-2-T', 'NSP14-280-0-T', 'NSP15-176-2-T', 'NC-26491-T',\n",
    "                   'NC-28271-T'])\n",
    "\n",
    "all_variants   = [epsilon, alpha, beta, gamma, lambda_new, delta, twentyE_EU1, D614G, omicron, ba2, ba5, ba4, ba2121]\n",
    "np.save(os.path.join(DATA_DIR, 'variants.npy'), all_variants)\n",
    "\n",
    "var_dic = {\n",
    "    'alpha' : alpha, \n",
    "    'epsilon' : epsilon, \n",
    "    'beta' : beta, \n",
    "    'gamma' : gamma, \n",
    "    'lambda' : lambda_new, \n",
    "    'delta' : delta, \n",
    "    '20E\\(EU1\\)' : twentyE_EU1, \n",
    "    'B.1' : D614G,\n",
    "    'omicron' : omicron,\n",
    "    'BA.2' : ba2,\n",
    "    'eta' : eta,\n",
    "    'iota' : iota,\n",
    "    'B.1.1.318' : B1_1_318,\n",
    "    'kappa' : kappa,\n",
    "    'mu' : mu,\n",
    "    'BA.4' : ba4,\n",
    "    'BA.5' : ba5,\n",
    "    'BA.2.12.1' : ba2121\n",
    "}\n",
    "\n",
    "variants_full = [var_dic[i] for i in var_dic]\n",
    "np.save(os.path.join(DATA_DIR, 'variants-all.npy'), variants_full)\n",
    "\n",
    "np.save(os.path.join(DATA_DIR, 'alpha-mutations.npy'),   alpha)\n",
    "np.save(os.path.join(DATA_DIR, 'gamma-mutations.npy'),   gamma)\n",
    "np.save(os.path.join(DATA_DIR, 'delta-mutations.npy'),   delta)\n",
    "np.save(os.path.join(DATA_DIR, 'omicron-mutations.npy'), omicron)\n",
    "\n",
    "# Saving the order in which the variants appeared for later early detection analysis\n",
    "group_order   = ['B.1', 'Alpha', 'Beta', 'Epsilon', 'Gamma', 'Eta', 'Iota', 'Kappa', 'Lambda', 'Mu', 'Delta', 'BA.1', 'BA.2', 'BA.2.12.1', 'BA.4', 'BA.5']\n",
    "variant_order = [[D614G], [alpha, beta, epsilon], [gamma, eta, iota, kappa, lambda_new], [mu], [delta], [omicron], [ba2], [ba2121, ba4, ba5]]\n",
    "\n",
    "new_groups = []\n",
    "old_muts   = []\n",
    "for i in range(len(variant_order)):\n",
    "    current_muts = []\n",
    "    for group in variant_order[i]:\n",
    "        new = [j for j in group if j not in old_muts]\n",
    "        new_groups.append(new)\n",
    "        for j in new:\n",
    "            current_muts.append(j)\n",
    "    for j in current_muts:\n",
    "        if j not in old_muts:\n",
    "            old_muts.append(j)\n",
    "            \n",
    "np.save(os.path.join(SARS_DIR, 'variants-exclusive.npy'), new_groups)\n",
    "\n",
    "# THE BELOW ARE INCOMPLETE LISTS OF MUTATIONS\n",
    "b12    = ['NSP5-89-0-T', 'NSP14-129-0-G', 'ORF3a-172-1-T', 'ORF8-24-1-T', 'N-67-0-T']    # this is B.1.2\n",
    "b11318 = ['NSP15-320-0-A', 'S-575-2-C', 'S-1238-2-A', 'ORF7b-44-2--', 'NC-27887--', 'NC-27888--', 'NC-27889--', 'NC-27890--',     # This is B.1.1.318, S:D796H\n",
    "          'NC-27891--', 'NC-27892--', 'ORF8-1-0--', 'ORF8-1-1--', 'ORF8-1-2--', 'NSP4-173-1-T', 'S-796-0-C', 'ORF8-2-0--', \n",
    "          'ORF8-2-1--', 'ORF8-2-2--', 'ORF8-3-0--', 'ORF8-3-1--', 'NC-28270-G', 'N-208-2--', 'N-209-0--', 'N-208-1--']\n",
    "b1429  = ['NSP8-3-2-T', 'NSP2-598-0-C', 'NSP4-131-2-T', 'NSP9-65-0-G', 'S-929-2-C', 'NC-27889-T', 'NSP2-530-2-T']\n",
    "\n",
    "\n",
    "# ADD ZETA, THETA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1524ccbc",
   "metadata": {},
   "source": [
    "<a id=\"4\"></a> \n",
    "## Data preprocessing\n",
    "\n",
    "#### extracting accessions, virus name, date, and location information from metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0314a3e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/metadata-2024-01-26 &&\n",
      "tar -xf /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/metadata_tsv_2023_06_05.tar.xz -C /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/metadata-2024-01-26\n"
     ]
    }
   ],
   "source": [
    "print(f'mkdir {METADATA_DIR} &&')\n",
    "print(f'tar -xf {METADATA_COMP} -C {METADATA_DIR}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6939c342",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extracting accessions\n",
    "\n",
    "script_file      = 'subset-metadata.py'\n",
    "\n",
    "job_file         = 'metaprocess.sh'\n",
    "job_str          = f\"\"\"#!/bin/bash -l\n",
    "\n",
    "#SBATCH --job-name=metaprocess\n",
    "#SBATCH --output=metaprocess.log\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --partition=big_memory\n",
    "#SBATCH --mem=32G\n",
    "#SBATCH --mail-user={USER_EMAIL}\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --time=6:00:00\n",
    "\n",
    "echo Running on `hostname`\n",
    "echo workdir $SLURM_SUBMIT_DIR\n",
    "\n",
    "#input_directory = \"{SSH_DATA}\"\n",
    "#cd input_directory\n",
    "\n",
    "cd $SLURM_SUBMIT_DIR\n",
    "\n",
    "#scratch drive folder\n",
    "SCRDIR=/scr/${SLURM_JOB_ID}\n",
    "\n",
    "if [[ ! -e $SCRDIR ]]; then\n",
    "\tmkdir $SCRDIR\n",
    "\n",
    "fi\n",
    "\n",
    "chmod +rX $SCRDIR\n",
    "\n",
    "echo scratch drive ${SCRDIR}\n",
    "\n",
    "cd $SCRDIR\n",
    "\n",
    "cp $SLURM_SUBMIT_DIR/subset-metadata.py ${SCRDIR}\n",
    "cp $SLURM_SUBMIT_DIR/metadata.tsv ${SCRDIR}\n",
    "\n",
    "\n",
    "# load and activate conda environment\n",
    "module load anaconda\n",
    "#conda activate align-env\n",
    "conda activate sars-env\n",
    "import pandas as pd \n",
    "\n",
    "# input and output files\n",
    "\n",
    "input_tsv=\"metadata.tsv\"\n",
    "output_tsv=\"outputmeta.tsv\"\n",
    "output_csv=\"outputmeta.csv\"\n",
    "\n",
    "#copy files back to working dir on exit\n",
    "trap \"mv outputmeta.tsv $SLURM_SUBMIT_DIR\" EXIT\n",
    "trap \"mv outputmeta.csv $SLURM_SUBMIT_DIR\" EXIT\n",
    "\n",
    "# Redirect stdout and stderr to the log file\n",
    "exec > metaprocess.log 2>&1\n",
    "\n",
    "# Print job information\n",
    "echo \"SLURM job ID: $SLURM_JOB_ID\"\n",
    "echo \"Start time: $(date)\"\n",
    "\n",
    "# Run the Python script\n",
    "python subset_metadata.py \"$input_tsv\" \"$output_tsv\"\n",
    "\n",
    "echo \"metaprocess completed!\"\n",
    "echo \"Saved to: $output_tsv\"\n",
    "\n",
    "csv_table=pd.read_table(output_tsv,sep='\\t')\n",
    "csv_table.to_csv('output_csv',index=False)\n",
    "\n",
    "# Print job completion information\n",
    "echo \"conversion completed!\"\n",
    "echo \"Saved to: $output_csv\"\n",
    "\"\"\"\n",
    "\n",
    "f = open(os.path.join(SARS_DIR, job_file), mode='w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "print('scp %s/%s %s '   % (SARS_DIR, job_file, SSH_HOME))\n",
    "print('')\n",
    "\n",
    "# commands to execute on the cluster\n",
    "\n",
    "print('sbatch %s' % job_file)\n",
    "print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4a83a49",
   "metadata": {},
   "source": [
    "#### Un-tar file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56ebd863",
   "metadata": {},
   "outputs": [],
   "source": [
    "msa_file2 = f'{msa_name}.tar.xz'\n",
    "msa_path2 = os.path.join(SARS_DIR, msa_file2)\n",
    "tar -xf {SSH_DATA}/{msa_file2} -C {SSH_DATA}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "daa54685",
   "metadata": {},
   "source": [
    "#### merging sequence data with metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8be74090",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scp /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/data-processing-pipeline/msa_2023-05-29.fasta efinn@cluster.csb.pitt.edu:/net/dali/home/barton/efinn/SC2/2023-06-05 &&\n",
      "scp /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/data-processing-pipeline/outputmeta.csv efinn@cluster.csb.pitt.edu:/net/dali/home/barton/efinn/SC2/2023-06-05\n",
      "scp /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/data-processing-pipeline/job-merge-db.sh efinn@cluster.csb.pitt.edu: \n",
      "\n",
      "sbatch job-merge-db.sh\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Merging sequence data in GISAID alignment with metadata on the cluster\n",
    "\n",
    "# Write sequences to lightning-mapped database and extract location information from metadata\n",
    "\n",
    "msa_name         = MSA_NAME\n",
    "msa_file         = f'{msa_name}.fasta'\n",
    "msa_path_cluster = ppath.join(SSH_DATA, msa_name, msa_name + '.fasta')\n",
    "script_file      = 'merge-metadata-lmdb-parallel.py'\n",
    "\n",
    "num_dirs  = int(15600000 / 500000) + 1\n",
    "\n",
    "job_file         = 'job-merge-db.sh'\n",
    "job_str          = f\"\"\"#!/bin/bash -l\n",
    "\n",
    "#SBATCH --time=7-00:00:00\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=256G\n",
    "#SBATCH --job-name=merge\n",
    "#SBATCH -e ./MPL/out/merge-error-%a\n",
    "#SBATCH -o ./MPL/out/merge-out-%a\n",
    "#SBATCH --mail-user={USER_EMAIL}\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --array=0-{num_dirs}\n",
    "\n",
    "# load and activate conda environment\n",
    "module load anaconda\n",
    "conda activate align-env\n",
    "\n",
    "python {script_file} --meta_file {os.path.join(SSH_DATA, METADATA_INTERIM)} --msa_file {msa_path_cluster} --lmdbDir {DATABASE_DIR} --refFile {REF_FILE} --dbNumber $SLURM_ARRAY_TASK_ID --metaNew {METADATA_NEW}\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SARS_DIR, job_file), mode='w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "meta_path      = os.path.join(SARS_DIR, METADATA_INTERIM)\n",
    "msa_path_local = os.path.join(SARS_DIR, msa_file)\n",
    "print(f'scp {msa_path_local} {SSH_HOME}{SSH_DATA} &&')\n",
    "print(f'scp {meta_path} {SSH_HOME}{SSH_DATA}')\n",
    "print('scp %s/%s %s '   % (SARS_DIR, job_file, SSH_HOME))\n",
    "print('')\n",
    "\n",
    "# commands to execute on the cluster\n",
    "\n",
    "print('sbatch %s' % job_file)\n",
    "print('')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "191b449a",
   "metadata": {},
   "source": [
    "#### find gaps, subset regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b26d8377",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scp /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/data-processing/gap-list.npy efinn@cluster.csb.pitt.edu:\n"
     ]
    }
   ],
   "source": [
    "# find list of known gaps\n",
    "\n",
    "ref_df    = pd.read_csv(os.path.join(DATA_DIR, f'ref-index-{DATA_DATE}.csv'))\n",
    "ref_index = list(ref_df['ref_index'])\n",
    "ref_sites = [dp.get_label_new(str(i) + '-A')[:-2] for i in ref_index]\n",
    "variants  = [epsilon, alpha, beta, gamma, lambda_new, delta, twentyE_EU1, D614G, omicron, ba2, ba5, ba4, ba2121, kappa, mu, b12, b11318, b1429, B1_1_318]\n",
    "gap_sites = []\n",
    "for var in variants:\n",
    "    for site in var:\n",
    "        if site[-1]=='-':\n",
    "            gap_sites.append(ref_index[ref_sites.index(site[:-2])])\n",
    "for i in list('abcdefghijklmnopqrstuvwxyz'):\n",
    "    gap_sites.append(f'22203{i}')\n",
    "out_file = os.path.join(DATA_DIR, 'gap-list.npy')\n",
    "np.save(out_file, gap_sites)\n",
    "\n",
    "# transfer list of known gaps to the cluster\n",
    "\n",
    "print(f'scp {out_file} {SSH_HOME}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7f5975c0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scp /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/data-processing-pipeline/SC2_BP_Data/processing-scripts/check-regional.py efinn001@cluster.hpcc.ucr.edu:/rhome/efinn001/shared/bigdata/deconvolution-2023-06-01/\n",
      "scp /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/data-processing-pipeline/SC2_BP_Data/processing-scripts/job-check-regional.sh efinn001@cluster.hpcc.ucr.edu:/rhome/efinn001/shared/bigdata/deconvolution-2023-06-01/\n",
      "\n",
      "sbatch job-check-regional.sh\n",
      "\n",
      "scp efinn001@cluster.hpcc.ucr.edu:/rhome/efinn001/bigdata/SARS-CoV-2-Data/2023-06-05/regional-2023-06-05.csv /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/data-processing-pipeline/SC2_BP_Data\n"
     ]
    }
   ],
   "source": [
    "# finding the number of sequences in each region\n",
    "out_file      = f'regional-{DATA_DATE}'\n",
    "script_file   = 'check-regional.py'\n",
    "job_dir       = SSH_HOME + SSH_DATA\n",
    "job_file      = 'job-check-regional.sh'\n",
    "job_str       = f\"\"\"#!/bin/bash -l\n",
    "\n",
    "#SBATCH --time=0-3\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem-per-cpu=100G\n",
    "#SBATCH --job-name=regional-check\n",
    "#SBATCH --mail-user={USER_EMAIL}\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -e ./region-error\n",
    "#SBATCH -o ./region-out\n",
    "#SBATCH -p batch\n",
    "\n",
    "python {script_file} --alignment {os.path.join(SSH_DATA, METADATA_INTERIM)} -o {os.path.join(SSH_DATA, out_file)}\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), mode='w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "# transfer data processing scripts and job file to the cluster\n",
    "\n",
    "script_path = os.path.join(SCRIPT_DIR, script_file)\n",
    "job_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {script_path} {SSH_HOME} {SSH_DATA}')\n",
    "print(f'scp {job_path} {SSH_HOME} {SSH_DATA}')\n",
    "print('')\n",
    "\n",
    "# run the job on the cluster\n",
    "\n",
    "print('sbatch %s' % job_file)\n",
    "print('')\n",
    "\n",
    "# transfer output data back to local directory\n",
    "\n",
    "out_path = os.path.join(job_dir, out_file + '.csv')\n",
    "print('scp %s %s' % (out_path, SARS_DIR))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "54057fe5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Regions that are selected manually because either\n",
    "# 1. They are broken up in an abnormal manner\n",
    "# 2. For some of the pandemic the case numbers in the region were too low for the model to be reliable \n",
    "#(or there was essentially no community transmission)\n",
    "\n",
    "north_cali = ['santa clara county, alameda county', 'san francisco county', 'sacramento county', \n",
    "              'contra costa county', 'humboldt county', 'sonoma county', 'stanislaus county', \n",
    "              'san joaquin county', 'solano county', 'merced county', 'santa cruz',\n",
    "              'madera county', 'southern san joaquin valley', 'placer county',  \n",
    "              'davis', 'nevada county', 'santa cruz county', 'san mateo county', 'butte county', \n",
    "              'mendocino county', 'shasta county', 'tuolumne county', 'napa county',\n",
    "              'marin county', 'el dorado county', 'lake county', 'san luis obispo', \n",
    "              'sutter county', 'contra costa', 'tehama county']\n",
    "\n",
    "south_cali = ['los angeles', 'los angeles county', 'orange', 'orange county', 'san diego', \n",
    "              'san diego county', 'ventura', 'ventura county', 'san luis obispo', \n",
    "              'san luis obispo county', 'imperial', 'imperial county', 'san bernardino county', \n",
    "              'riverside county', 'kern county', 'imperial county']\n",
    "\n",
    "regs = [['europe', 'united kingdom', 'northern ireland', None, None, None],\n",
    "        ['europe', 'united kingdom', ['england', 'wales', 'scotland'], None, None, None],\n",
    "        ['north america', 'usa', 'california', south_cali, None, None],\n",
    "        ['north america', 'usa', 'california', north_cali, None, None]]\n",
    "\n",
    "state_elim   = ['california']\n",
    "country_elim = []\n",
    "for entry in regs:\n",
    "    if entry[1]!='usa':\n",
    "        country_elim.append(entry[1])\n",
    "\n",
    "# Large countries or states, that must be broken up into smaller regions\n",
    "\n",
    "use_subregions  = [          'usa',        'mexico',        'canada', 'turkey', 'india',        'brazil', 'russia',     'argentina',         'chile']\n",
    "cont_subregions = ['north america', 'north america', 'north america',   'asia',  'asia', 'south america', 'europe', 'south america', 'south america']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1d26d7e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(os.path.join(SARS_DIR, f'regional-{DATA_DATE}.csv'))\n",
    "continent = [str(i) for i in list(df['continent'])]\n",
    "country   = [str(i) for i in list(df['country'])]\n",
    "combined  = [continent[i] + '-' + country[i] for i in range(len(continent))]\n",
    "df['combined'] = combined\n",
    "data           = df['combined'].value_counts()\n",
    "\n",
    "regions = np.array(list(data.axes[0]))\n",
    "counts  = np.array(list(data))\n",
    "mask    = counts > MIN_REGIONAL_SEQS\n",
    "regions = regions[mask]\n",
    "counts  = counts[mask]\n",
    "\n",
    "regions    = [i.split('-') for i in regions]\n",
    "continents = [i[0] for i in regions]\n",
    "countries  = [i[1] for i in regions]\n",
    "\n",
    "for i in range(len(regions)):\n",
    "    if countries[i] not in use_subregions and countries[i] not in country_elim:\n",
    "        regs.append([continents[i], countries[i], None, None, None, None])\n",
    "\n",
    "for i in range(len(use_subregions)):\n",
    "    continent = cont_subregions[i]\n",
    "    country   = use_subregions[i]\n",
    "    df_temp   = df[(df['continent']==continent) & (df['country']==country)]\n",
    "    data      = df_temp['region'].value_counts()\n",
    "    states    = np.array(list(data.axes[0]))\n",
    "    counts    = np.array(list(data))\n",
    "    \n",
    "    mask      = counts > 1000\n",
    "    states    = states[mask]\n",
    "    counts    = counts[mask]\n",
    "    \n",
    "    for j in range(len(states)):\n",
    "        if states[j] not in state_elim:\n",
    "            regs.append([continent, country, states[j], None, None, None]) \n",
    "\n",
    "regs.sort(key = lambda x: x[:2])      \n",
    "selected = regs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "638741ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "297"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(selected)\n",
    "\n",
    "# not shown: manual sub-setting of 297 regions for sample set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e1abedc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "regions_dir = os.path.join(SARS_DIR, f'regions-{DATA_DATE}')\n",
    "if not os.path.exists(regions_dir):\n",
    "    os.mkdir(regions_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4f1b8f9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "for group in selected:\n",
    "    new = np.array(group)[np.array(group)!=None]\n",
    "    contains_list = False\n",
    "    for i in new:\n",
    "        if not isinstance(i, str):\n",
    "            contains_list = True\n",
    "    if contains_list:\n",
    "        label = ''\n",
    "        for i in new:\n",
    "            if isinstance(i, str):\n",
    "                label += i\n",
    "            else:\n",
    "                if len(i) > 4:\n",
    "                    #print(i)\n",
    "                    if new[2]=='california':\n",
    "                        if 'riverside county' in new[3]:\n",
    "                            label += 'south'\n",
    "                        else:\n",
    "                            label += 'north'\n",
    "                else:\n",
    "                    label += '_'.join(i)\n",
    "            if i!=new[-1]:\n",
    "                label += '-'\n",
    "    else:\n",
    "        label = '-'.join(new)\n",
    "    f = open(os.path.join(regions_dir, label + '.npy'), mode='wb')\n",
    "    np.save(f, [group])\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "1f1850aa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scp /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/processing-scripts/find-region-times-parallel.py efinn@cluster.csb.pitt.edu:/net/dali/home/barton/efinn/deconvolution-new &&\n",
      "scp /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/processing-scripts/job-find-regions-parallel.sh efinn@cluster.csb.pitt.edu:/net/dali/home/barton/efinn/deconvolution-new &&\n",
      "scp -r /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/data-processing/regions-2024-01-26-sub efinn@cluster.csb.pitt.edu:/net/dali/home/barton/efinn/deconvolution-new\n",
      "\n",
      "sbatch job-find-regions-parallel.sh\n"
     ]
    }
   ],
   "source": [
    "# Transfer scripts, run processing, and extract files\n",
    "\n",
    "# Processing clips the time series in different regions according to set rules \n",
    "    # regarding the distribution of the number of sampled genomes over time\n",
    "\n",
    "regions_dir_sub = os.path.join(DATA_DIR, f'regions-{DATA_DATE}-sub')\n",
    "out_folder    = ppath.join(SSH_DATA, 'regions-times')\n",
    "script_file   = 'find-region-times-parallel.py'\n",
    "region_data_sub   = f'regions-{DATA_DATE}-sub'\n",
    "\n",
    "job_file      = 'job-find-regions-parallel.sh'\n",
    "job_str       = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=50G\n",
    "#SBATCH --time=1-00:00:00\n",
    "#SBATCH --mail-user={USER_EMAIL}\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -e ./logs/region-error-%a\n",
    "#SBATCH -o ./logs/region-out-%a\n",
    "#SBATCH --array=0-29\n",
    "\n",
    "regions=({region_data_sub}/*)\n",
    "region=${{regions[$SLURM_ARRAY_TASK_ID]}}\n",
    "python {script_file} --input_file {ppath.join(SSH_DATA, METADATA_INT)} -o {out_folder} --regions \\\"$region\\\" \n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), 'w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "job_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "script_path = os.path.join(SCRIPT_DIR, script_file)\n",
    "print(f'scp {script_path} {SSH_HOME}{SSH_DATA} &&')\n",
    "print(f'scp {job_path} {SSH_HOME}{SSH_DATA} &&')\n",
    "print('scp -r %s %s/net/dali/home/barton/efinn/deconvolution-new' % (regions_dir_sub, SSH_HOME))\n",
    "print('')\n",
    "\n",
    "print('sbatch %s' % job_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff7d867f",
   "metadata": {},
   "source": [
    "<a id=\"5\"></a> \n",
    "## Processing\n",
    "#### Extracting sequences linked to metadata, removing extreme genome gaps, imputation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "85697d8d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ls /net/dali/home/barton/efinn/deconvolution-new/regions-times | wc -l &&\n",
      "\n",
      "scp /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/processing-scripts/extract-seqs.py efinn@cluster.csb.pitt.edu:/net/dali/home/barton/efinn/deconvolution-new &&\n",
      "scp /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/processing-scripts/job-extract.sh efinn@cluster.csb.pitt.edu:/net/dali/home/barton/efinn/deconvolution-new\n",
      "\n",
      "sbatch job-extract.sh\n"
     ]
    }
   ],
   "source": [
    "# Processing with multiple lightning mapped databases storing the sequences\n",
    "\n",
    "out_folder    = ppath.join(SSH_DATA, 'msa-reg')\n",
    "script_file   = 'extract-seqs.py'\n",
    "region_dir    = ppath.join(SSH_DATA, f'regions-times')\n",
    "lmdb_dir      = SSH_LMDB\n",
    "\n",
    "# run on cluster\n",
    "\n",
    "print(f'ls {region_dir} | wc -l &&')\n",
    "print('')\n",
    "num_regions = 32\n",
    "\n",
    "job_file      = 'job-extract.sh'\n",
    "job_str       = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --mail-user={USER_EMAIL}\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -e ./logs/genome-error-%a\n",
    "#SBATCH -o ./logs/genome-out-%a\n",
    "#SBATCH --array=0-{num_regions - 1}\n",
    "\n",
    "module purge\n",
    "module load anaconda\n",
    "eval \"$(conda shell.bash hook)\"\n",
    "conda activate align-env\n",
    "\n",
    "regions=({region_dir}/*)\n",
    "region=${{regions[$SLURM_ARRAY_TASK_ID]}}\n",
    "python {script_file} --input_file {METADATA_NEW} -o {out_folder} --regions \\\"$region\\\"  --lmdb {lmdb_dir}\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), 'w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "# transfer job file to cluster\n",
    "\n",
    "script_path = os.path.join(SCRIPT_DIR, script_file)\n",
    "job_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {script_path} {SSH_HOME}{SSH_DATA} &&')\n",
    "print(f'scp {job_path} {SSH_HOME}{SSH_DATA}')\n",
    "print('')\n",
    "\n",
    "print('sbatch %s' % job_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "94896181",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ls /net/dali/home/barton/efinn/deconvolution-new/msa-reg | wc -l &&\n",
      "\n",
      "scp /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/processing-scripts/job-process.sh efinn@cluster.csb.pitt.edu:\n",
      "\n",
      "sbatch job-process.sh\n"
     ]
    }
   ],
   "source": [
    "# Additional processing and removing excess gaps\n",
    "\n",
    "out_folder    = ppath.join(SSH_DATA, 'genome-data')\n",
    "input_dir     = ppath.join(SSH_DATA, 'msa-reg')\n",
    "script_file   = 'process-threaded.py'\n",
    "threads       = 20\n",
    "\n",
    "print(f'ls {input_dir} | wc -l &&')\n",
    "print('')\n",
    "num_regions = 30\n",
    "\n",
    "job_file      = 'job-process.sh'\n",
    "job_str       = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --cpus-per-task={threads}\n",
    "#SBATCH --mem=150G\n",
    "#SBATCH --mail-user={USER_EMAIL}\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -e ./logs/genome-error-%a\n",
    "#SBATCH -o ./logs/genome-out-%a\n",
    "#SBATCH --array=0-{num_regions - 1}\n",
    "\n",
    "files=({input_dir}/*)\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "python {script_file} --input_file \\\"$file\\\" -o {out_folder} --refFile {REF_FILE + '.csv'} --threads {threads}\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), 'w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "script_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {script_path} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "print('sbatch %s' % job_file)\n",
    "\n",
    "#scp /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/data-processing/ref-index-2024-01-26.csv efinn@cluster.csb.pitt.edu:/net/dali/home/barton/efinn/deconvolution-new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9d33dff8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "readarray -d '' files < <(find /net/dali/home/barton/efinn/deconvolution-new/genome-data ! -name \"*sites*\" -print0) &&\n",
      "echo ${#files[*]}\n",
      "\n",
      "scp /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/processing-scripts/job-mut-counts.sh efinn@cluster.csb.pitt.edu:/net/dali/home/barton/efinn/deconvolution-new &&\n",
      "\n",
      "scp /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/processing-scripts/find-mut-counts.py efinn@cluster.csb.pitt.edu:/net/dali/home/barton/efinn/deconvolution-new\n",
      "\n",
      "sbatch job-mut-counts.sh\n"
     ]
    }
   ],
   "source": [
    "# Finding counts and imputing gaps\n",
    "\n",
    "input_dir   = ppath.join(SSH_DATA, 'genome-data')\n",
    "mut_count_dir     = ppath.join(SSH_DATA, 'mutant-counts')\n",
    "\n",
    "mut_script    = 'find-mut-counts.py'\n",
    "script_file = mut_script\n",
    "\n",
    "\n",
    "# run on cluster to find number of files\n",
    "\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0) &&\"\"\")    # in order to find the number of files in the input directory\n",
    "print('echo ${#files[*]}')\n",
    "print('')\n",
    "\n",
    "num_files     = 31   # the number of data files \n",
    "\n",
    "job_file      = 'job-mut-counts.sh'\n",
    "job_str       = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --partition=big_memory\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --mail-user={USER_EMAIL}\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -e ./logs/mut-counts-error-%a\n",
    "#SBATCH -o ./logs/mut-counts-out-%a\n",
    "#SBATCH --array=1-{num_files-1}\n",
    "\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "\n",
    "module load anaconda\n",
    "\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "python {mut_script} --input_file \\\"$file\\\" -o {mut_count_dir} --refFile {REF_FILE + '.csv'}\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), 'w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "job_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {job_path} {SSH_HOME}{SSH_DATA} &&')\n",
    "print('')\n",
    "\n",
    "script_path = os.path.join(SCRIPT_DIR, script_file)\n",
    "print(f'scp {script_path} {SSH_HOME}{SSH_DATA}')\n",
    "print('')\n",
    "\n",
    "print('sbatch %s' % job_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c24adbbb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "readarray -d '' files < <(find /net/dali/home/barton/efinn/deconvolution-new/genome-data ! -name \"*sites*\" -print0) &&\n",
      "echo ${#files[*]}\n",
      "\n",
      "scp /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/processing-scripts/job-impute.sh efinn@cluster.csb.pitt.edu:/net/dali/home/barton/efinn/deconvolution-new &&\n",
      "\n",
      "scp /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/processing-scripts/impute-gaps.py efinn@cluster.csb.pitt.edu:/net/dali/home/barton/efinn/deconvolution-new\n",
      "\n",
      "sbatch job-impute.sh\n"
     ]
    }
   ],
   "source": [
    "# imputing gaps\n",
    "\n",
    "input_dir   = ppath.join(SSH_DATA, 'genome-data')\n",
    "mut_count_dir     = ppath.join(SSH_DATA, 'mutant-counts')\n",
    "impute_dir = ppath.join(SSH_DATA, 'impute-gaps')\n",
    "\n",
    "impute_script    = 'impute-gaps.py'\n",
    "script_file = impute_script\n",
    "\n",
    "\n",
    "# run on cluster to find number of files\n",
    "\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0) &&\"\"\")\n",
    "print('echo ${#files[*]}')\n",
    "print('')\n",
    "\n",
    "num_files     = 31   # the number of data files \n",
    "\n",
    "job_file      = 'job-impute.sh'\n",
    "job_str       = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --partition=big_memory\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --mail-user={USER_EMAIL}\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -e ./logs/impute-error-%a\n",
    "#SBATCH -o ./logs/impute-out-%a\n",
    "#SBATCH --array=1-{num_files-1}\n",
    "\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "\n",
    "module load anaconda\n",
    "\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "python {impute_script} --input \\\"$file\\\" -o {impute_dir} --mut_dir {mut_count_dir} --refFile {REF_FILE + '.csv'}\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), 'w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "job_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {job_path} {SSH_HOME}{SSH_DATA} &&')\n",
    "print('')\n",
    "\n",
    "script_path = os.path.join(SCRIPT_DIR, script_file)\n",
    "print(f'scp {script_path} {SSH_HOME}{SSH_DATA}')\n",
    "print('')\n",
    "\n",
    "print('sbatch %s' % job_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "176634b8",
   "metadata": {},
   "source": [
    "<a id=\"6\"></a>\n",
    "## Back Projection Processing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a28b89a",
   "metadata": {},
   "source": [
    "<a id=\"7\"></a>\n",
    "### Step 1: take formatted (subprocessed) sequence file, divide into unique and non-unique sequences for back projection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45017ce3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# divide into unique sequences (only one instance) and non-unique (multiple counts)\n",
    "\n",
    "input_dir = ppath.join(SSH_DATA, 'impute-gaps')\n",
    "unique_dir     = ppath.join(SSH_DATA, 'unique')\n",
    "unique_script    = 'unique.py'\n",
    "script_file = unique_script\n",
    "\n",
    "\n",
    "# run on cluster to find number of \n",
    "\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0) &&\"\"\")\n",
    "print('echo ${#files[*]}')\n",
    "print('')\n",
    "\n",
    "num_files     = 31   # the number of data files \n",
    "\n",
    "job_file      = 'job-unique.sh'\n",
    "\n",
    "job_str       = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --partition=big_memory\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --mail-user={USER_EMAIL}\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -e ./logs/unique-error-%a\n",
    "#SBATCH -o ./logs/unique-out-%a\n",
    "#SBATCH --array=1-{num_files-1}\n",
    "\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "\n",
    "module load anaconda\n",
    "\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "python {unique_script} --input \\\"$file\\\" -o {unique_dir}\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), 'w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "job_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {job_path} {SSH_HOME}{SSH_DATA} &&')\n",
    "print('')\n",
    "\n",
    "script_path = os.path.join(SCRIPT_DIR, script_file)\n",
    "print(f'scp {script_path} {SSH_HOME}{SSH_DATA}')\n",
    "print('')\n",
    "\n",
    "print('sbatch %s' % job_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "643e7394",
   "metadata": {},
   "source": [
    "<a id=\"8\"></a>\n",
    "### step 2: transfer bp files to cluster, run deconvolution (back-projection) script"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bff66cb",
   "metadata": {},
   "source": [
    "### Step 2a: setting up submission directory to run deconvolution (back-projection) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "cbb9f2bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# country_list.txt is a file with manually selected regions\n",
    "\n",
    "os.chdir(SCRIPT_DIR)\n",
    "with open('country_list.txt', 'r') as f:\n",
    "    with open('country_list_command.txt', 'a') as w_f:\n",
    "        for line in f:\n",
    "            country = line.strip()\n",
    "            command = f'ls {SSH_DATA}/genome-unique/{country}/output-{country} | wc -l'\n",
    "            w_f.write(command + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "ba8f1a74",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sbatch job-bp-ems-africa-cameroon.sh &&\n",
      "sbatch job-bp-ems-africa-kenya.sh &&\n",
      "sbatch job-bp-ems-africa-reunion.sh &&\n",
      "sbatch job-bp-ems-africa-south-africa-2020-01-02-2022-11-25.sh &&\n",
      "sbatch job-bp-ems-africa-south-africa-2022-11-25-2024-01-25.sh &&\n",
      "sbatch job-bp-ems-asia-hong-kong.sh &&\n",
      "sbatch job-bp-ems-asia-india-gujarat.sh &&\n",
      "sbatch job-bp-ems-asia-india-karnataka.sh &&\n",
      "sbatch job-bp-ems-asia-indonesia-2020-01-02-2023-03-31.sh &&\n",
      "sbatch job-bp-ems-asia-indonesia-2023-03-31-2024-01-30.sh &&\n",
      "sbatch job-bp-ems-asia-vietnam.sh &&\n",
      "sbatch job-bp-ems-europe-bosnia-and-herzegovina.sh &&\n",
      "sbatch job-bp-ems-europe-croatia.sh &&\n",
      "sbatch job-bp-ems-europe-cyprus.sh &&\n",
      "sbatch job-bp-ems-europe-estonia.sh &&\n",
      "sbatch job-bp-ems-europe-latvia.sh &&\n",
      "sbatch job-bp-ems-europe-portugal.sh &&\n",
      "sbatch job-bp-ems-europe-romania.sh &&\n",
      "sbatch job-bp-ems-europe-russia-moscow.sh &&\n",
      "sbatch job-bp-ems-europe-slovakia.sh &&\n",
      "sbatch job-bp-ems-north-america-canada-manitoba.sh &&\n",
      "sbatch job-bp-ems-north-america-canada-quebec-2020-01-02-2022-08-08.sh &&\n",
      "sbatch job-bp-ems-north-america-canada-quebec-2022-08-08-2024-01-22.sh &&\n",
      "sbatch job-bp-ems-north-america-costa-rica.sh &&\n",
      "sbatch job-bp-ems-north-america-jamaica.sh &&\n",
      "sbatch job-bp-ems-north-america-mexico-mexico-city.sh &&\n",
      "sbatch job-bp-ems-north-america-mexico-nuevo-leon.sh &&\n",
      "sbatch job-bp-ems-north-america-panama.sh &&\n",
      "sbatch job-bp-ems-north-america-puerto-rico.sh &&\n",
      "sbatch job-bp-ems-north-america-usa-wyoming.sh &&\n",
      "scp -r /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/processing-scripts/bp-submit-dir efinn@cluster.csb.pitt.edu:/net/dali/home/barton/efinn/deconvolution-new &&\n",
      "scp /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/processing-scripts/back_projection_ems_parallel.py efinn@cluster.csb.pitt.edu:/net/dali/home/barton/efinn/deconvolution-new\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv('countries_sh_maker.csv')\n",
    "\n",
    "for index, row in df.iterrows():\n",
    "    country = row['country']\n",
    "    num_files = row['num_files'] #1638\n",
    "\n",
    "\n",
    "    script_file = 'back_projection_ems_parallel.py'\n",
    "    job_file = f'job-bp-ems-{country}.sh'\n",
    "    job_str = f\"\"\"#!/bin/bash -l\n",
    "\n",
    "#SBATCH --job-name=back-projection-{country}\n",
    "#SBATCH --output=./bp_logs/bp_{country}_%A_%a.out\n",
    "#SBATCH --error=./bp_logs/bp_{country}_%A_%a.err\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=2G\n",
    "#SBATCH --mail-user={USER_EMAIL}\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --array=0-{num_files}\n",
    "\n",
    "echo Running on `hostname`\n",
    "echo workdir $SLURM_SUBMIT_DIR\n",
    "\n",
    "# Load necessary modules and activate environment\n",
    "module load anaconda\n",
    "conda activate align-env\n",
    "\n",
    "# Define input and output directories\n",
    "input_dir='{SSH_DATA}/genome-unique/{country}/output-{country}'\n",
    "output_dir='{SSH_DATA}/genome-unique/{country}/bp-output-{country}'\n",
    "\n",
    "# Create output directory if it does not exist\n",
    "if [[ ! -d $output_dir ]]; then\n",
    "mkdir -p $output_dir\n",
    "fi\n",
    "\n",
    "# Define the list of input files\n",
    "input_files=(\"$input_dir\"/*.csv)\n",
    "\n",
    "# Get the input file for this job based on SLURM array task ID\n",
    "input_file=\"${{input_files[$SLURM_ARRAY_TASK_ID]}}\"\n",
    "\n",
    "# Define scratch directory\n",
    "SCRDIR=/scr/${{SLURM_JOB_ID}}\n",
    "\n",
    "# Create scratch directory\n",
    "if [[ ! -e $SCRDIR ]]; then\n",
    "    mkdir $SCRDIR\n",
    "fi\n",
    "\n",
    "chmod +rX $SCRDIR\n",
    "\n",
    "echo scratch drive ${{SCRDIR}}\n",
    "\n",
    "cd $SCRDIR\n",
    "\n",
    "# Copy input files to scratch directory\n",
    "cp \"$input_file\" $SCRDIR\n",
    "cp $SLURM_SUBMIT_DIR/back_projection_ems_parallel.py ${{SCRDIR}}\n",
    "cp $SLURM_SUBMIT_DIR/incubation_period_sars2.csv ${{SCRDIR}}\n",
    "cp $SLURM_SUBMIT_DIR/weights_binom.csv ${{SCRDIR}}\n",
    "\n",
    "# Run the script from the scratch directory\n",
    "cd $SCRDIR\n",
    "\n",
    "# Run the script with the correct read directory\n",
    "python back_projection_ems_parallel.py -input_file \"$(basename \"$input_file\")\" -write_dir \"$output_dir\" -incubation_file 'incubation_period_sars2.csv' -smoothing_file 'weights_binom.csv'\n",
    "\n",
    "# Print job completion information\n",
    "echo \"SLURM job ID: $SLURM_JOB_ID\"\n",
    "echo \"Array task ID: $SLURM_ARRAY_TASK_ID\"\n",
    "echo \"Job completed at $(date)\"\n",
    "\"\"\"\n",
    "    \n",
    "    with open(os.path.join(SCRIPT_DIR, 'bp-submit-dir', job_file), 'w') as f:\n",
    "        f.write(job_str)\n",
    "    \n",
    "    print('sbatch %s &&' % job_file)\n",
    "        \n",
    "print('scp -r %s/%s %s%s &&' % (SCRIPT_DIR, 'bp-submit-dir', SSH_HOME, SSH_DATA))\n",
    "print('scp %s/%s %s%s' % (SCRIPT_DIR, script_file, SSH_HOME, SSH_DATA))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97046049",
   "metadata": {},
   "source": [
    "<a id=\"9\"></a>\n",
    "### Step 3: reformatting to run downstream steps"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfd3894d",
   "metadata": {},
   "source": [
    "### Step 3a: create format directory for slurm scripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "90d9658a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sbatch job-format-africa-cameroon.sh &&\n",
      "sbatch job-format-africa-kenya.sh &&\n",
      "sbatch job-format-africa-reunion.sh &&\n",
      "sbatch job-format-africa-south-africa-2020-01-02-2022-11-25.sh &&\n",
      "sbatch job-format-africa-south-africa-2022-11-25-2024-01-25.sh &&\n",
      "sbatch job-format-asia-hong-kong.sh &&\n",
      "sbatch job-format-asia-india-gujarat.sh &&\n",
      "sbatch job-format-asia-india-karnataka.sh &&\n",
      "sbatch job-format-asia-indonesia-2020-01-02-2023-03-31.sh &&\n",
      "sbatch job-format-asia-indonesia-2023-03-31-2024-01-30.sh &&\n",
      "sbatch job-format-asia-vietnam.sh &&\n",
      "sbatch job-format-europe-bosnia-and-herzegovina.sh &&\n",
      "sbatch job-format-europe-croatia.sh &&\n",
      "sbatch job-format-europe-cyprus.sh &&\n",
      "sbatch job-format-europe-estonia.sh &&\n",
      "sbatch job-format-europe-latvia.sh &&\n",
      "sbatch job-format-europe-portugal.sh &&\n",
      "sbatch job-format-europe-romania.sh &&\n",
      "sbatch job-format-europe-russia-moscow.sh &&\n",
      "sbatch job-format-europe-slovakia.sh &&\n",
      "sbatch job-format-north-america-canada-manitoba.sh &&\n",
      "sbatch job-format-north-america-canada-quebec-2020-01-02-2022-08-08.sh &&\n",
      "sbatch job-format-north-america-canada-quebec-2022-08-08-2024-01-22.sh &&\n",
      "sbatch job-format-north-america-costa-rica.sh &&\n",
      "sbatch job-format-north-america-jamaica.sh &&\n",
      "sbatch job-format-north-america-mexico-mexico-city.sh &&\n",
      "sbatch job-format-north-america-mexico-nuevo-leon.sh &&\n",
      "sbatch job-format-north-america-panama.sh &&\n",
      "sbatch job-format-north-america-puerto-rico.sh &&\n",
      "sbatch job-format-north-america-usa-wyoming.sh &&\n",
      "scp -r /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/processing-scripts/bp-submit-dir efinn@cluster.csb.pitt.edu:/net/dali/home/barton/efinn/deconvolution-new &&\n",
      "scp /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/processing-scripts/format-deconvolution.py efinn@cluster.csb.pitt.edu:/net/dali/home/barton/efinn/deconvolution-new\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv('countries_sh_maker.csv')\n",
    "\n",
    "for index, row in df.iterrows():\n",
    "    country = row['country']\n",
    "    \n",
    "    script_file = 'format-deconvolution.py'\n",
    "    job_file = f'job-format-{country}.sh'\n",
    "    job_str = f\"\"\"#!/bin/bash\n",
    "\n",
    "#SBATCH --job-name=bp-format-{country}\n",
    "#SBATCH --output=./format_logs/format_{country}.out\n",
    "#SBATCH --error=./format_logs/format_{country}.err\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=64G\n",
    "#SBATCH --mail-user={USER_EMAIL}\n",
    "#SBATCH --mail-type=ALL\n",
    "\n",
    "module load anaconda\n",
    "conda activate align_env\n",
    "\n",
    "output_dir='{SSH_DATA}/bp-formatted/'\n",
    "input_dir='{SSH_DATA}'\n",
    "sites_dir='{SSH_DATA}/impute-gaps/'\n",
    "location='{country}'\n",
    "\n",
    "python format-deconvolution.py -o \"$output_dir\" --input_dir \"$input_dir\" --sites_dir \"$sites_dir\" --location \"$location\" --prob_dist 'single-sequence-distribution.csv'\n",
    "\"\"\"\n",
    "    \n",
    "    with open(os.path.join(SCRIPT_DIR, 'bp-format-sh-dir', job_file), 'w') as f:\n",
    "        f.write(job_str)\n",
    "    \n",
    "    print('sbatch %s &&' % job_file)\n",
    "        \n",
    "print('scp -r %s/%s %s%s &&' % (SCRIPT_DIR, 'bp-format-sh-dir', SSH_HOME, SSH_DATA))\n",
    "print('scp %s/%s %s%s' % (SCRIPT_DIR, script_file, SSH_HOME, SSH_DATA))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3536958",
   "metadata": {},
   "source": [
    "<a id=\"10\"></a>\n",
    "# trim regions, filter sites\n",
    "#### after formatting: trimming, filtering for sites of interest (frequencies above 5%)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c69b0d16",
   "metadata": {},
   "source": [
    "<a id=\"11\"></a>\n",
    "### trim back projected regions, individual jobs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "42be38a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sbatch job-trim-africa-cameroon.sh &&\n",
      "sbatch job-trim-africa-kenya.sh &&\n",
      "sbatch job-trim-africa-reunion.sh &&\n",
      "sbatch job-trim-africa-south-africa-2020-01-02-2022-11-25.sh &&\n",
      "sbatch job-trim-africa-south-africa-2022-11-25-2024-01-25.sh &&\n",
      "sbatch job-trim-asia-hong-kong.sh &&\n",
      "sbatch job-trim-asia-india-gujarat.sh &&\n",
      "sbatch job-trim-asia-india-karnataka.sh &&\n",
      "sbatch job-trim-asia-indonesia-2020-01-02-2023-03-31.sh &&\n",
      "sbatch job-trim-asia-indonesia-2023-03-31-2024-01-30.sh &&\n",
      "sbatch job-trim-asia-vietnam.sh &&\n",
      "sbatch job-trim-europe-bosnia-and-herzegovina.sh &&\n",
      "sbatch job-trim-europe-croatia.sh &&\n",
      "sbatch job-trim-europe-cyprus.sh &&\n",
      "sbatch job-trim-europe-estonia.sh &&\n",
      "sbatch job-trim-europe-latvia.sh &&\n",
      "sbatch job-trim-europe-portugal.sh &&\n",
      "sbatch job-trim-europe-romania.sh &&\n",
      "sbatch job-trim-europe-russia-moscow.sh &&\n",
      "sbatch job-trim-europe-slovakia.sh &&\n",
      "sbatch job-trim-north-america-canada-manitoba.sh &&\n",
      "sbatch job-trim-north-america-canada-quebec-2020-01-02-2022-08-08.sh &&\n",
      "sbatch job-trim-north-america-canada-quebec-2022-08-08-2024-01-22.sh &&\n",
      "sbatch job-trim-north-america-costa-rica.sh &&\n",
      "sbatch job-trim-north-america-jamaica.sh &&\n",
      "sbatch job-trim-north-america-mexico-mexico-city.sh &&\n",
      "sbatch job-trim-north-america-mexico-nuevo-leon.sh &&\n",
      "sbatch job-trim-north-america-panama.sh &&\n",
      "sbatch job-trim-north-america-puerto-rico.sh &&\n",
      "sbatch job-trim-north-america-usa-wyoming.sh &&\n",
      "scp -r /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/processing-scripts/bp-trim-sh-dir efinn@cluster.csb.pitt.edu:/net/dali/home/barton/efinn/deconvolution-new &&\n",
      "scp /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/processing-scripts/trimming_bp.py efinn@cluster.csb.pitt.edu:/net/dali/home/barton/efinn/deconvolution-new\n"
     ]
    }
   ],
   "source": [
    "# creating sh files from list of countries\n",
    "\n",
    "df = pd.read_csv('countries_sh_maker.csv')\n",
    "\n",
    "for index, row in df.iterrows():\n",
    "    country = row['country']\n",
    "    \n",
    "    script_file = 'trimming-bp.py'\n",
    "    job_file = f'job-trim-{country}.sh'\n",
    "    job_str = f\"\"\"#!/bin/bash\n",
    "\n",
    "#SBATCH --job-name=bp-trim-{country}\n",
    "#SBATCH --output=./trim_logs/trim_{country}.out\n",
    "#SBATCH --error=./trim_logs/trim_{country}.err\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=64G\n",
    "#SBATCH --mail-user={USER_EMAIL}\n",
    "#SBATCH --mail-type=ALL\n",
    "\n",
    "module load anaconda\n",
    "conda activate align_env\n",
    "\n",
    "input_dir='{SSH_DATA}/bp-formatted/'\n",
    "output_dir='{SSH_DATA}/bp-trim-20/'\n",
    "location='{country}'\n",
    "\n",
    "python trimming-bp.py --input_dir \"$input_dir\" -o \"$output_dir\" --location \"$location\" -x 20 --thresh 0.04 -n 5\n",
    "\"\"\"\n",
    "    \n",
    "    with open(os.path.join(SCRIPT_DIR, 'bp-trim-sh-dir', job_file), 'w') as f:\n",
    "        f.write(job_str)\n",
    "    \n",
    "    print('sbatch %s &&' % job_file)\n",
    "        \n",
    "print('scp -r %s/%s %s%s &&' % (SCRIPT_DIR, 'bp-trim-sh-dir', SSH_HOME, SSH_DATA))\n",
    "print('scp %s/%s %s%s' % (SCRIPT_DIR, script_file, SSH_HOME, SSH_DATA))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c8b591b",
   "metadata": {},
   "source": [
    "<a id=\"12\"></a>\n",
    "### trimming naive\n",
    "#### for comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ac388ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# trimming step, naive: finding time series with good sampling\n",
    "\n",
    "input_dir     = ppath.join(SSH_DATA, 'genome-data')\n",
    "output_dir    = ppath.join(SSH_DATA)\n",
    "script_file   = 'trim-intervals-sd.py'\n",
    "\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\"\"\")\n",
    "print('echo ${#files[*]}')\n",
    "print('')\n",
    "\n",
    "job_file      = 'job-trim-intervals.sh'\n",
    "job_str       = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --partition=big_memory\n",
    "#SBATCH -e ./sampling-error-%a\n",
    "#SBATCH -o ./sampling-out-%a\n",
    "#SBATCH --array=1-5\n",
    "\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "\n",
    "python {script_file} --input \\\"$file\\\" -o {output_dir} --trimDir genome-trimmed\n",
    "\"\"\" \n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), 'w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "script_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {script_path} {SSH_HOME}')\n",
    "print('')\n",
    "\n",
    "print('sbatch %s' % job_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1eb84a07",
   "metadata": {},
   "source": [
    "<a id=\"13\"></a>\n",
    "### finding max frequencies and filtering\n",
    "#### for comparison with back projection"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad3b2604",
   "metadata": {},
   "source": [
    "<a id=\"14\"></a>\n",
    "### step 1: find naive max freq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdff7576",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find max, naive\n",
    "\n",
    "input_dir = ppath.join(SSH_DATA, 'genome-trimmed')\n",
    "max_dir = ppath.join(SSH_DATA, 'genome-max-freqs')\n",
    "\n",
    "max_freq_script = 'find-max-frequencies.py'\n",
    "script_file = max_freq_script\n",
    "\n",
    "# run on cluster to find number of files\n",
    "\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0) &&\"\"\")\n",
    "print('echo ${#files[*]}')\n",
    "print('')\n",
    "\n",
    "num_files     = 180   # the number of data files \n",
    "\n",
    "job_file      = 'job-max-freq.sh'\n",
    "\n",
    "job_str       = f\"\"\"#!/bin/bash -l\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH -e ./logs/freqs-error-%a\n",
    "#SBATCH -o ./logs/freqs-out-%a\n",
    "#SBATCH --mail-user={USER_EMAIL}\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --array=1-{num_files-1}\n",
    "\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "\n",
    "file=${files[$SLURM_ARRAY_TASK_ID]}\n",
    "python {max_freq_script} --input_file \"$file\" -o {max_dir} --refFile {REF_FILE + '.csv'}\n",
    "\"\"\"\n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), 'w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "job_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {job_path} {SSH_HOME}{SSH_DATA} &&')\n",
    "print('')\n",
    "\n",
    "script_path = os.path.join(SCRIPT_DIR, script_file)\n",
    "print(f'scp {script_path} {SSH_HOME}{SSH_DATA}')\n",
    "print('')\n",
    "\n",
    "print('sbatch %s' % job_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc43f099",
   "metadata": {},
   "source": [
    "<a id=\"15\"></a>\n",
    "### step 2 find masks and filter naive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cece4412",
   "metadata": {},
   "outputs": [],
   "source": [
    "# individual masks: mask off sites for time-regions that rise above 5% frequency in naive method\n",
    "\n",
    "input_dir = ppath.join(SSH_DATA, 'genome-trimmed')\n",
    "filtered_dir = ppath.join(SSH_DATA, 'genome-filtered-05')\n",
    "max_dir = ppath.join(SSH_DATA, 'genome-max-freqs')\n",
    "max_mask_dir = ppath.join(SSH_DATA, 'genome-max-mask-05')\n",
    "\n",
    "filter_mask_script    = 'filter-sites-mask.py'\n",
    "script_file = filter_mask_script\n",
    "\n",
    "# run on cluster to find number of files\n",
    "\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0) &&\"\"\")\n",
    "print('echo ${#files[*]}')\n",
    "print('')\n",
    "\n",
    "num_files     = 180   # the number of data files \n",
    "\n",
    "job_file      = 'job-filter-sites-mask-05.sh'\n",
    "\n",
    "job_str       = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=56G\n",
    "#SBATCH -e ./logs/mask-error-%a\n",
    "#SBATCH -o ./logs/mask-out-%a\n",
    "#SBATCH --mail-user={USER_EMAIL}\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --array=1-{num_files-1}\n",
    "\n",
    "module load anaconda\n",
    "\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "file=${files[$SLURM_ARRAY_TASK_ID]}\n",
    "python {filter_mask_script} --input \"$file\" -o {filtered_dir} --max_dir {max_dir} --mask_dir {max_mask_dir} --refFile {REF_FILE + '.csv'} --min_freq 0.05\n",
    "\"\"\"\n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), 'w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "job_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {job_path} {SSH_HOME}{SSH_DATA} &&')\n",
    "print('')\n",
    "\n",
    "script_path = os.path.join(SCRIPT_DIR, script_file)\n",
    "print(f'scp {script_path} {SSH_HOME}{SSH_DATA}')\n",
    "print('')\n",
    "\n",
    "print('sbatch %s' % job_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0eaae114",
   "metadata": {},
   "source": [
    "<a id=\"16\"></a>\n",
    "### step 3: combine masks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92fd3f33",
   "metadata": {},
   "outputs": [],
   "source": [
    "# combined masks (incomplete; run locally) \n",
    "# TO DO: make py script\n",
    "# creates a common mask for time-regions in each region\n",
    "\n",
    "def get_region_name(file_name):\n",
    "    # extract region filename\n",
    "    return file_name.split('---')[0]\n",
    "\n",
    "def print_true_indices(mask, label):\n",
    "    indices = np.argwhere(mask)\n",
    "    print(f\"\\n{label} true indices:\")\n",
    "    for index in indices:\n",
    "        print(tuple(index))\n",
    "\n",
    "def load_combine_masks(file_list, region):\n",
    "    combined_mask = None\n",
    "    for file in file_list:\n",
    "        mask = np.load(file)\n",
    "        # print (optional for debug) true indices for individual mask\n",
    "        print_true_indices(mask, f\"{region} - {file}\")\n",
    "        if combined_mask is None:\n",
    "            combined_mask = mask\n",
    "        else:\n",
    "            combined_mask = np.logical_or(combined_mask, mask)\n",
    "    return combined_mask\n",
    "\n",
    "# main function to run in notebook\n",
    "def combine_main(directory):\n",
    "    mask_files = []\n",
    "    \n",
    "    # directory walk for mask files\n",
    "    for root, dirs, files in os.walk(directory):\n",
    "        for file in files:\n",
    "            if file.endswith('-mask.npy'):\n",
    "                mask_files.append(os.path.join(root, file))\n",
    "    \n",
    "    # group region files\n",
    "    region_files = defaultdict(list)\n",
    "    for file in mask_files:\n",
    "        region = get_region_name(os.path.basename(file))\n",
    "        region_files[region].append(file)\n",
    "    \n",
    "    # combine mask\n",
    "    combined_masks = {}\n",
    "    for region, files in region_files.items():\n",
    "        combined_mask = load_combine_masks(files, region)\n",
    "        combined_masks[region] = combined_mask\n",
    "        # print (optional, remove if not needed)\n",
    "        print_true_indices(combined_mask, f\"{region} - combined mask\")\n",
    "    \n",
    "    # save combined mask\n",
    "    for region, mask in combined_masks.items():\n",
    "        output_file = os.path.join(directory, f\"{region}-combined-mask.npy\")\n",
    "        np.save(output_file, mask)\n",
    "        print(f\"Combined mask saved to: {output_file}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f72c4642",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run for combined masks to get comparison frequency files\n",
    "\n",
    "input_dir = ppath.join(SSH_DATA, 'genome-trimmed')\n",
    "filter_comparison_dir = ppath.join(SSH_DATA, 'genome-filtered-compare-05')\n",
    "combine_mask_dir = ppath.join(SSH_DATA, 'genome-max-combine-mask-05')\n",
    "    # combine mask directory made locally and uploaded from previous cell\n",
    "\n",
    "filter_combine_script    = 'filter-sites-combine-mask.py'\n",
    "script_file = filter_combine_mask_script\n",
    "\n",
    "# run on cluster to find number of \n",
    "\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0) &&\"\"\")\n",
    "print('echo ${#files[*]}')\n",
    "print('')\n",
    "\n",
    "num_files     = 180   # the number of data files \n",
    "\n",
    "job_file      = 'job-filter-sites-compare-mask-05.sh'\n",
    "job_str       = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=32G\n",
    "#SBATCH -e ./logs/mask-combine-error-%a\n",
    "#SBATCH -o ./logs/mask-combine-out-%a\n",
    "#SBATCH --mail-user={USER_EMAIL}\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --array=1-{num_files-1}\n",
    "\n",
    "module load anaconda\n",
    "\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "\n",
    "python {filter_combine_script} --input \"$file\" -o {filter_comparison_dir} --mask_dir {combine_mask_dir} --refFile {REF_FILE + '.csv'}\n",
    "\"\"\"\n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), 'w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "job_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {job_path} {SSH_HOME}{SSH_DATA} &&')\n",
    "print('')\n",
    "\n",
    "script_path = os.path.join(SCRIPT_DIR, script_file)\n",
    "print(f'scp {script_path} {SSH_HOME}{SSH_DATA}')\n",
    "print('')\n",
    "\n",
    "print('sbatch %s' % job_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "138cfe5c",
   "metadata": {},
   "source": [
    "<a id=\"17\"></a>\n",
    "### step 4: subset back projection site files with previously made masks "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd5a1545",
   "metadata": {},
   "outputs": [],
   "source": [
    "# individual masks\n",
    "\n",
    "input_dir   = ppath.join(SSH_DATA, 'bp-trim-20')\n",
    "bp_filter_dir = ppath.join(SSH_DATA, 'bp-filter-20-05')\n",
    "max_mask_dir = ppath.join(SSH_DATA, 'genome-max-mask-05')\n",
    "\n",
    "bp_filter_script    = 'filter-sites-bp-mask.py'\n",
    "script_file = bp_filter_script\n",
    "\n",
    "# run on cluster to find number of files\n",
    "\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0) &&\"\"\")\n",
    "print('echo ${#files[*]}')\n",
    "print('')\n",
    "\n",
    "num_files      = 180\n",
    "\n",
    "job_file       = 'job-filter-bp-sites-mask-05.sh'\n",
    "job_str        = f\"\"\"#!/bin/bash \n",
    "#SBATCH --job-name=filt-bp-mask\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --mem=50G\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mail-user={USER_EMAIL}\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -e ./logs/bp-filt-error-%a\n",
    "#SBATCH -o ./logs/bp-filt-out-%a\n",
    "#SBATCH --array=1-{num_files-1}\n",
    "\n",
    "\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "\n",
    "python {bp_filter_script} -o {bp_filter_dir} --input \"$file\" --mask_dir {max_mask_dir} --refFile {REF_FILE + '.csv'}\n",
    "\"\"\"\n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), 'w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "job_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {job_path} {SSH_HOME}{SSH_DATA} &&')\n",
    "print('')\n",
    "\n",
    "script_path = os.path.join(SCRIPT_DIR, script_file)\n",
    "print(f'scp {script_path} {SSH_HOME}{SSH_DATA}')\n",
    "print('')\n",
    "\n",
    "print('sbatch %s' % job_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6119548c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# combined masks\n",
    "\n",
    "input_dir   = ppath.join(SSH_DATA, 'bp-trim-20')\n",
    "bp_filter_compare_dir = ppath.join(SSH_DATA, 'bp-filtered-compare-05')\n",
    "combine_mask_dir = ppath.join(SSH_DATA, 'genome-max-combine-mask-05')\n",
    "\n",
    "bp_filter_combine_script    = 'filter-sites-bp-combine-mask.py'\n",
    "script_file = bp_filter_script\n",
    "\n",
    "# run on cluster to find number of files\n",
    "\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0) &&\"\"\")\n",
    "print('echo ${#files[*]}')\n",
    "print('')\n",
    "\n",
    "num_files      = 149\n",
    "\n",
    "job_file       = 'job-filter-bp-sites-compare-mask-05.sh'\n",
    "job_str        = f\"\"\"#!/bin/bash -l\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=32G\n",
    "#SBATCH -e ./logs/bp-mask-combine-error-%a\n",
    "#SBATCH -o ./logs/bp-mask-combine-out-%a\n",
    "#SBATCH --mail-user={USER_EMAIL}\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH --array=1-{num_files-1}\n",
    "\n",
    "\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "\n",
    "python {bp_filter_combine_script} -o {bp_filter_compare_dir} --input \"$file\" --mask_dir {combine_mask_dir} --refFile {REF_FILE + '.csv'}\n",
    "\"\"\"\n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), 'w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "job_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {job_path} {SSH_HOME}{SSH_DATA} &&')\n",
    "print('')\n",
    "\n",
    "script_path = os.path.join(SCRIPT_DIR, script_file)\n",
    "print(f'scp {script_path} {SSH_HOME}{SSH_DATA}')\n",
    "print('')\n",
    "\n",
    "print('sbatch %s' % job_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d8e9715",
   "metadata": {},
   "source": [
    "<a id=\"18\"></a>\n",
    "### Covariance and Inference\n",
    "\n",
    "This section is a bit ad hoc."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f15ac1cb",
   "metadata": {},
   "source": [
    "<a id=\"19\"></a>\n",
    "### covariance, cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4ef8c82",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir = ppath.join(SSH_DATA, 'bp-filtered-compare-05')\n",
    "temp_dir = ppath.join(SSH_DATA, 'bp-covar')\n",
    "scratch_dir =  ppath.join(SSH_DATA, 'scratch-bp-covar')\n",
    "\n",
    "covar_bp_script    = 'epi-covar-bp-cluster.py'\n",
    "script_file = covar_bp_script\n",
    "\n",
    "# run on cluster to find number of files\n",
    "\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0) &&\"\"\")\n",
    "print('echo ${#files[*]}')\n",
    "print('')\n",
    "\n",
    "num_files      = 149\n",
    "\n",
    "job_file       = 'job-epi-covar-bp.sh'\n",
    "\n",
    "job_str        = f\"\"\"#!/bin/bash -l\n",
    "\n",
    "#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --mail-user={USER_EMAIL}\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -e ./logs/mpl/bp-freq-error-%a\n",
    "#SBATCH -o ./logs/mpl/bp-freq-out-%a\n",
    "#SBATCH --array=1-{num_files-1}%1\n",
    "\n",
    "cd ./inference-c++\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "g++ src/main-N.cpp src/inf.cpp src/io.cpp -O3 -march=native -lgslcblas -lgsl -o bin/mpl -std=c++11\n",
    "\n",
    "file=${{files[$SLURM_ARRAY_TASK_ID]}}\n",
    "tempout={temp_dir}\n",
    "python ../{covar_bp_script} --data \"$file\" -o $tempout -q 5 --pop_size 10000 -k 0.1 -R 2 --scratch {scratch_dir} --timed 1\n",
    "\"\"\"\n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), 'w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "job_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {job_path} {SSH_HOME}{SSH_DATA} &&')\n",
    "print('')\n",
    "\n",
    "script_path = os.path.join(SCRIPT_DIR, script_file)\n",
    "print(f'scp {script_path} {SSH_HOME}{SSH_DATA}')\n",
    "print('')\n",
    "\n",
    "print('sbatch %s' % job_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c44cb7e1",
   "metadata": {},
   "source": [
    "#### covariance, naive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9229f244",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir = ppath.join(SSH_DATA, 'genome-filtered-compare-05')\n",
    "temp_dir = ppath.join(SSH_DATA, 'genome-covar')\n",
    "scratch_dir =  ppath.join(SSH_DATA, 'scratch-genome-covar')\n",
    "\n",
    "covar_script    = 'epi-covar-cluster.py'\n",
    "script_file = covar_script\n",
    "\n",
    "# run on cluster to find number of files\n",
    "\n",
    "print(f\"\"\"readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0) &&\"\"\")\n",
    "print('echo ${#files[*]}')\n",
    "print('')\n",
    "\n",
    "num_files      = 180\n",
    "\n",
    "job_file       = 'job-epi-covar.sh'\n",
    "\n",
    "job_str        = f\"\"\"#!/bin/bash -l \n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=1\n",
    "#SBATCH --mem=100G\n",
    "#SBATCH --mail-user={USER_EMAIL}\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -e ./logs/mpl/freq-error-%a\n",
    "#SBATCH -o ./logs/mpl/freq-out-%a\n",
    "#SBATCH --array=1-{num_files-1}%1\n",
    "\n",
    "cd ./inference-c++\n",
    "readarray -d '' files < <(find {input_dir} ! -name \"*sites*\" -print0)\n",
    "g++ src/main-N.cpp src/inf.cpp src/io.cpp -O3 -march=native -lgslcblas -lgsl -o bin/mpl -std=c++11\n",
    "\n",
    "file=${files[$SLURM_ARRAY_TASK_ID]}\n",
    "tempout={temp_dir}\n",
    "python ../{covar_script} --data \"$file\" -o $tempout -q 5 --pop_size 10000 -k 0.1 -R 2 --scratch {scratch_dir} --timed 1\n",
    "\"\"\"\n",
    "\n",
    "f = open(os.path.join(SCRIPT_DIR, job_file), 'w')\n",
    "f.write('%s\\n' % job_str)\n",
    "f.close()\n",
    "\n",
    "job_path = os.path.join(SCRIPT_DIR, job_file)\n",
    "print(f'scp {job_path} {SSH_HOME}{SSH_DATA} &&')\n",
    "print('')\n",
    "\n",
    "script_path = os.path.join(SCRIPT_DIR, script_file)\n",
    "print(f'scp {script_path} {SSH_HOME}{SSH_DATA}')\n",
    "print('')\n",
    "\n",
    "print('sbatch %s' % job_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8345517e",
   "metadata": {},
   "source": [
    "<a id=\"20\"></a>\n",
    "### inference: local running on cluster files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d28d90a",
   "metadata": {},
   "outputs": [],
   "source": [
    "inf_script    = 'epi-inf-parallel.py'\n",
    "archive_loc   = os.path.join(DATA_DIR, 'inference-c++')\n",
    "covar_home    = os.path.join(DATA_DIR, 'bp-covar-south-africa')\n",
    "\n",
    "os.chdir(DATA_DIR)\n",
    "\n",
    "covar_country_target = os.path.join(covar_home, 'africa-south-africa')\n",
    "\n",
    "for i in os.listdir(covar_country_target):\n",
    "    if i == \".DS_Store\":\n",
    "            continue\n",
    "    covar_dir      = os.path.join(covar_country_target, f'{i}')\n",
    "    out_file       = os.path.join(home_dir, 'inference-south-africa', f'{i}-inf-out')\n",
    "    %run {inf_script} --data {covar_dir} --timed 1 -o {out_file} -q 5 --g1 40 --saveCovar --refFile {REF_FILE}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afc7c43a",
   "metadata": {},
   "source": [
    "<a id=\"21\"></a>\n",
    "### local running, cluster script not supported"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9d167b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# FOR MACS\n",
    "\n",
    "homebrew run command:  g++ src/main-N.cpp src/inf.cpp src/io.cpp -O3 -march=native -lgslcblas -lgsl -L/opt/homebrew/Cellar/gsl/2.7.1/lib -o bin/mpl -std=c++11"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a0b3fb6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# infering the selection coefficients using a cutoff frequency of 5%\n",
    "import os\n",
    "\n",
    "freq_script = 'epi-covar-bp.py'\n",
    "inf_script  = 'epi-inf-parallel.py'\n",
    "\n",
    "filter_input_file = 'Europe-Portugal---2021-4-20-2022-11-6.csv'\n",
    "\n",
    "\n",
    "out_inf_dir   = f'{DATA_DIR}/inf-bp-out'\n",
    "#in_filt_dir   = f'{DATA_DIR}/filt-bp-20'\n",
    "\n",
    "\n",
    "os.chdir(PROCESS_DIR)\n",
    "\n",
    "archive_loc    = os.path.join(INF_DIR)\n",
    "temp_dir       = os.path.join(PROCESS_DIR, 'temp')\n",
    "out_file       = os.path.join(out_inf_dir, f'infer-Europe-Portugal---2021-4-20-2022-11-6.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "81928b27",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scratch directory:    /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/data-processing-pipeline/temp\n",
      "covariance directory: /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/data-processing-pipeline/temp/temp-sd-covar-dir\n",
      "outfile /Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/data-processing-pipeline/temp\n",
      "sequence file\t Europe-Portugal---2021-4-20-2022-11-6.csv\n",
      "sequence lengths are [76]\n",
      "number of sequences with each length [129586]\n",
      "[  685   686   687   688   689   690   691   692   693  6512  6513  6514\n",
      " 11282 11283 11284 11285 11286 11287 11288 11289 11290 11291 11292 11293\n",
      " 11294 11295 21632 21633 21634 21635 21636 21637 21638 21639 21640 21764\n",
      " 21765 21766 21767 21768 21769 21986 21987 21988 21989 21990 21991 21992\n",
      " 21993 21994 22028 22029 22030 22031 22032 22033 22193 22194 22195 27888\n",
      " 28247 28248 28249 28250 28251 28252 28270 28361 28362 28363 28364 28365\n",
      " 28366 28367 28368 28369]\n",
      "number of sites:\t76\n",
      "        times    counts                                          sequences\n",
      "0         475  0.082220  [1, 1, 3, 4, 2, 1, 4, 4, 4, 3, 4, 4, 3, 4, 4, ...\n",
      "1         475  0.082220  [1, 1, 3, 4, 2, 1, 4, 4, 4, 3, 4, 4, 3, 4, 4, ...\n",
      "2         475  0.082220  [1, 1, 3, 4, 2, 1, 4, 4, 4, 3, 4, 4, 3, 4, 4, ...\n",
      "3         475  0.237397  [1, 1, 3, 4, 2, 1, 4, 4, 4, 3, 4, 4, 3, 4, 4, ...\n",
      "16        475  0.132691  [1, 1, 3, 4, 2, 1, 4, 4, 4, 3, 4, 4, 3, 4, 4, ...\n",
      "...       ...       ...                                                ...\n",
      "129416   1040  0.137993  [1, 1, 3, 4, 2, 1, 4, 4, 4, 3, 4, 4, 3, 4, 4, ...\n",
      "127955   1040  0.064613  [1, 1, 3, 4, 2, 1, 4, 4, 4, 3, 4, 4, 3, 4, 4, ...\n",
      "85562    1040  0.047892  [1, 1, 3, 4, 2, 1, 4, 4, 4, 3, 4, 4, 3, 4, 4, ...\n",
      "129572   1040  0.063987  [1, 1, 3, 4, 2, 1, 4, 4, 4, 3, 4, 4, 3, 4, 4, ...\n",
      "129585   1040  0.303144  [1, 1, 3, 4, 2, 1, 4, 4, 4, 3, 4, 4, 3, 4, 4, ...\n",
      "\n",
      "[129586 rows x 3 columns]\n",
      "<_io.TextIOWrapper name='/Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/data-processing-pipeline/temp/temp-sd-covar-dir/seqs-Europe-Portugal---2021-4-20-2022-11-6.dat' mode='w' encoding='UTF-8'>\n",
      "<Popen: returncode: None args: ['/Users/liz/Documents/back_projection_home/b...>\n",
      "/Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/data-processing-pipeline/temp/temp-sd-covar-dir/covar-Europe-Portugal---2021-4-20-2022-11-6.dat\n",
      "/Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/data-processing-pipeline/temp/temp-sd-covar-dir/covar-Europe-Portugal---2021-4-20-2022-11-6.dat\n",
      "[  685   686   687   688   689   690   691   692   693  6512  6513  6514\n",
      " 11282 11283 11284 11285 11286 11287 11288 11289 11290 11291 11292 11293\n",
      " 11294 11295 21632 21633 21634 21635 21636 21637 21638 21639 21640 21764\n",
      " 21765 21766 21767 21768 21769 21986 21987 21988 21989 21990 21991 21992\n",
      " 21993 21994 22028 22029 22030 22031 22032 22033 22193 22194 22195 27888\n",
      " 28247 28248 28249 28250 28251 28252 28270 28361 28362 28363 28364 28365\n",
      " 28366 28367 28368 28369]\n",
      "[[0.         1.98048089 0.         ... 0.         0.         0.        ]\n",
      " [0.         2.26856071 0.         ... 0.         0.         0.        ]\n",
      " [0.         3.1718767  0.         ... 0.         0.         0.        ]\n",
      " ...\n",
      " [0.         3.70849365 0.         ... 0.         0.         0.36662599]\n",
      " [0.         1.28523519 0.         ... 0.         0.         0.08415413]\n",
      " [0.         0.90423828 0.         ... 0.         0.         0.0498824 ]]\n",
      "counts matrix: [[0.         1.98048089 0.         ... 0.         0.         0.        ]\n",
      " [0.         2.26856071 0.         ... 0.         0.         0.        ]\n",
      " [0.         3.1718767  0.         ... 0.         0.         0.        ]\n",
      " ...\n",
      " [0.         3.70849365 0.         ... 0.         0.         0.36662599]\n",
      " [0.         1.28523519 0.         ... 0.         0.         0.08415413]\n",
      " [0.         0.90423828 0.         ... 0.         0.         0.0498824 ]]\n",
      "[ 150.5165477   172.41061385  241.06262932  469.94520181  968.42795459\n",
      " 1891.9150029  3232.02858386 4688.71971461 5833.9256104  6296.909614\n",
      " 6062.99758651 5439.22122586 4724.53975514 4092.27554846 3504.10053103\n",
      " 2740.74514558 1754.0372366   876.37869302  377.63580305  242.14874705\n",
      "  246.88903693  291.37645301  347.59902936  368.13081174  360.75808726\n",
      "  359.14832835  383.14619521  428.25632728  483.85627206  518.27544445\n",
      "  508.84648883  438.08092483  376.91499998  381.28247057  590.95278703\n",
      " 1089.37760441 1970.85438385 3162.22183203 4424.89609678 5469.9343675\n",
      " 6126.97186465 6485.43278198 6653.7772858  6678.45462206 6552.08459517\n",
      " 6268.36939655 5717.75375473 4970.06088042 4244.12971459 3623.55389072\n",
      " 2871.60855056 1856.1519753  1114.5011247   951.31097959 1496.41386204\n",
      " 2416.00612538 3283.79516775 3627.17427649 3169.44678624 2240.02942731\n",
      " 1485.70244669 1170.49455521 1236.31799663 1409.71464793 1679.54641557\n",
      " 2208.75834508 3022.1902498  4141.18061081 5236.53778724 5845.74019503\n",
      " 5642.84408981 4714.18218686 3811.31561554 3709.20750653 4897.77723261\n",
      " 6786.42899033 8127.1274502  8035.85233101 6532.41119732 4863.51552205\n",
      " 4262.78930692 5064.95086075 6364.90406131 6860.20100719 6118.43459657\n",
      " 4569.06471976 3306.49106749 2987.24874363 3773.4165641  5235.32055435\n",
      " 6375.36390705 6394.78531216 5127.33618972 3357.02510192 2273.30398495\n",
      " 2391.96497172 3299.52387377 4141.26961564 4234.22257825 3502.19968968\n",
      " 2661.50898013 2568.00083171 3538.97217916 5082.77606413 6090.72304127\n",
      " 5882.84232434 4422.47830081 2528.76573087 1158.70341634  552.96893985\n",
      "  525.57875312  648.00613228  723.73751397  673.57746366  576.48844961\n",
      "  601.95861636  920.74445159 1534.83354164 2180.80359174 2489.67394238\n",
      " 2310.01926638 1973.06423955 2099.64716209 3183.18589776 5038.74087957\n",
      " 6722.18255089 7100.26238874 5795.17291192 3620.47942432 1993.90062718\n",
      " 1524.25387647 1877.23100826 2355.39709461 2541.58781897 2194.18560492\n",
      " 1572.36508566 1285.87393362 1804.67650724 2978.34306227 4089.26882736\n",
      " 4366.33109989 3616.18036778 2448.7090688  1746.2030742  1945.47013063\n",
      " 2683.08595664 3429.02967215 3681.12842124 3282.86930229 2619.58922249\n",
      " 2335.93536046 2823.99615077 3768.36862031 4578.51281115 4723.1675386\n",
      " 4118.36289712 3296.58950625 2993.27965972 3553.01057058 4482.4787732\n",
      " 5239.80051506 5160.72404512 4103.80393755 2492.54171523 1266.94376556\n",
      "  878.87777995  980.08292668 1165.9377345  1206.97695001 1131.94917645\n",
      " 1087.27245418 1384.99067417 2171.09670847 3244.12857169 4257.56894241\n",
      " 4627.8388223  4109.51885819 3064.37503662 2323.17821402 2335.94918989\n",
      " 2724.10482525 3066.52388123 3030.17144781 2666.07324268 2354.92591367\n",
      " 2544.11708591 3313.85786409 4152.55200404 4619.83112181 4594.92381482\n",
      " 4149.0343836  3466.45034007 2997.34636577 3280.12411048 4222.55593432\n",
      " 5239.30936644 5540.66200271 4781.91038917 3498.35727039 2703.31280589\n",
      " 2969.12764482 4051.37198787 5178.89930886 5648.03459378 5078.34000778\n",
      " 3932.66626214 3147.07174115 3401.62005317 4368.65809858 5239.50573116\n",
      " 5309.94156324 4463.24601556 3366.41012465 2851.95726571 3161.1237658\n",
      " 3744.14548844 3986.94688468 3655.18128205 3019.65356531 2728.38147044\n",
      " 3241.73919388 4463.81232896 5595.75114273 5889.15283275 5085.056253\n",
      " 3708.71473616 2536.24367918 2042.00975938 2395.95144845 3419.66367923\n",
      " 4507.59262874 4919.37118009 4243.02839062 2914.16035163 1868.0633591\n",
      " 1713.74908611 2370.56370455 3017.99726134 3086.90942485 2428.44359733\n",
      " 1562.17238426 1166.69764389 1565.42109092 2499.41651371 3345.84532991\n",
      " 3502.79463211 2817.62242422 1779.49439381 1152.71285401 1324.15541385\n",
      " 2128.80703387 2945.06490386 3182.92966665 2650.75838081 1727.03994578\n",
      " 1317.24529158 1806.895475   3030.36626848 4098.97235707 4266.17439461\n",
      " 3426.82776447 2218.71606567 1735.23022521 2297.46004287 3597.72826772\n",
      " 4742.67547758 4849.28729587 3798.14282049 2363.50888874 1651.8932532\n",
      " 2125.96123048 3454.59358437 4733.86778138 5028.50397471 3993.05761947\n",
      " 2376.11278392 1470.75054585 1956.32240096 3462.021308   4919.36148269\n",
      " 5233.21665803 4136.18622998 2499.68753347 1719.8108892  2451.50269853\n",
      " 4235.37261135 5977.58848394 6409.6488063  5117.88401985 3033.45005963\n",
      " 1733.96084763 1855.31275221 2939.36089892 3955.07215116 4059.09412912\n",
      " 3071.60884634 1708.52738431 1046.29371546 1437.37018078 2602.59580763\n",
      " 3736.88441245 4118.63650093 3517.95749106 2484.74152522 2125.31965644\n",
      " 2924.72561303 4643.60807739 6127.27223332 6337.5693158  5094.91149768\n",
      " 3311.82513928 2397.8702154  2850.80977364 4364.53897388 5891.22602918\n",
      " 6317.80766237 5166.25424661 3224.33781024 1808.67328159 1614.04452037\n",
      " 2286.20678816 3127.40425169 3467.77279347 3009.63204037 2211.98140124\n",
      " 1908.18092075 2748.22781866 4556.23712132 6375.59930236 6927.86450508\n",
      " 5680.92608008 3648.50992047 2383.40124169 2938.90676568 4756.4029358\n",
      " 6711.78464057 7450.64357982 6280.05504313 4273.82366325 2993.34143724\n",
      " 3530.30557243 5199.70881474 6748.53155577 7064.19919163 5731.45632478\n",
      " 3771.54680446 2588.88048386 3147.25079995 4816.10064028 6390.64197157\n",
      " 6722.23804887 5401.92953149 3548.94766249 2559.29535432 3296.47302879\n",
      " 4991.21433474 6417.82936211 6525.50821655 5083.30092649 3267.66974482\n",
      " 2375.52634897 3069.24273772 4854.12903899 6542.0063734  7133.56795353\n",
      " 6085.63660009 4183.61582759 3021.49768661 3581.51880317 5549.38516015\n",
      " 7319.8839326  7643.5962478  5982.60506826 3424.35508446 1426.15003371\n",
      "  792.16384958  935.51824031 1212.58241048 1308.15770968 1228.98704945\n",
      " 1317.25169908 2017.60754341 3621.4762853  5718.51578161 7286.08161783\n",
      " 7414.10135093 5946.69383671 4077.26688861 3148.74047318 3859.24657635\n",
      " 5599.86993816 7124.09876103 7337.10799686 5877.58288196 3920.14192834\n",
      " 2932.57978981 3751.05955631 5641.9819839  7175.29303176 7211.1440708\n",
      " 5503.96520209 3201.22660548 1702.88778908 1663.66772312 2707.43345938\n",
      " 3890.30859812 4377.93712622 3674.68505842 2476.01354101 1911.85639544\n",
      " 2600.34065324 3993.3468504  4902.62312897 4742.13344444 3573.65056427\n",
      " 2296.97014968 1787.14783141 2361.82805261 3602.60175207 4605.06469948\n",
      " 4850.89814689 4184.59012005 3228.45587987 2920.13323379 3795.37562888\n",
      " 5590.19514609 7267.38903796 8034.46978483 7324.21358889 5774.51848378\n",
      " 4545.19722669 4004.30061018 3982.32699761 3837.08913289 3548.7953201\n",
      " 2817.55011298 2083.89975957 1804.47112155 2026.88852933 2454.80544491\n",
      " 2670.32900606 2607.309499   2296.83480453 2043.02241046 2092.31340252\n",
      " 2451.08014535 2838.83884601 2965.38414395 2722.74808682 2145.53395541\n",
      " 1552.8163863  1238.89021531 1237.77434524 1299.63239753 1342.1676897\n",
      " 1484.8633327  1718.90197154 2011.42007493 2326.26723946 2552.90264198\n",
      " 2631.89211745 2525.59942019 2382.48512123 2167.31404045 1942.63859546\n",
      " 1824.18821352 1837.08633267 1967.69253808 2110.12015865 2239.87866559\n",
      " 2218.48829297 2117.32878017 2051.74275389 2090.56850772 2153.52659395\n",
      " 2133.3556566  2067.70569875 1936.24985288 1814.40569369 1854.83134955\n",
      " 2077.07429885 2388.00227254 2589.98595277 2683.77981497 2579.30900134\n",
      " 2410.61944291 2390.50595819 2518.03591556 2625.7604097  2553.44199634\n",
      " 2378.23205125 2091.80357532 1885.93518294 1927.78487632 2238.38804723\n",
      " 2569.55181559 2720.92048062 2620.07337673 2274.17475039 1946.24482047\n",
      " 1871.70654996 2022.07061227 2205.46852162 2304.78337772 2297.58926918\n",
      " 2122.40288992 1915.58038708 1907.4532051  2160.74564418 2444.15518164\n",
      " 2549.15705491 2483.31557842 2201.63407018 1983.48930778 1917.43217098\n",
      " 2091.96436063 2287.97489629 2356.10531558 2358.74099732 2212.5673448\n",
      " 2155.63805105 2225.79150208 2394.55205971 2454.5645808  2339.48141351\n",
      " 2230.35469224 2137.58983424 2138.7067095  2265.8844288  2326.43734626\n",
      " 2229.95000762 1949.74857339 1613.85704265 1179.26960454  693.01971744\n",
      "  319.32950326  176.52734036  181.72946871  250.65482416  459.54880058\n",
      "  808.79356938 1269.71848941 1689.96677086 1893.11880324 1875.98772832\n",
      " 1750.3432147  1571.01714087 1196.47655183  700.25828768  302.75979544\n",
      "  131.12502013  126.19318     199.07223581  381.14333405  733.20453508\n",
      " 1274.32772301 1839.74652269 2258.54880594 2359.50465996 2221.17148515\n",
      " 1885.82728616 1354.61562894  757.23729775  281.84551731   97.67787466\n",
      "   68.72210902]\n",
      "(566, 380)\n",
      "frequency matrix: [[0.         1.         0.         ... 0.         0.         0.        ]\n",
      " [0.         1.         0.         ... 0.         0.         0.        ]\n",
      " [0.         1.         0.         ... 0.         0.         0.        ]\n",
      " ...\n",
      " [0.         1.         0.         ... 0.         0.         0.09886116]\n",
      " [0.         1.         0.         ... 0.         0.         0.06547761]\n",
      " [0.         1.         0.         ... 0.         0.         0.05516511]]\n"
     ]
    }
   ],
   "source": [
    "%run epi-covar-bp --data {filter_input_file} -o {temp_dir} -q 5 --pop_size 10000 -k 0.1 -R 2 --scratch {temp_dir}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "01eeaa9a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "starting\n",
      "/Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/data-processing-pipeline/temp\n",
      ".DS_Store\n",
      "/Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/data-processing-pipeline/temp\n",
      "Europe-Portugal---2021-4-20-2022-11-6.npz\n",
      "/Users/liz/Documents/back_projection_home/backprojection-SARS-CoV-2-main/data-processing-pipeline/temp/Europe-Portugal---2021-4-20-2022-11-6.npz\n",
      "\tloading location Europe-Portugal---2021-4-20-2022-11-6\n",
      "[  685   686   687   688   689   690   691   692   693  6512  6513  6514\n",
      " 11282 11283 11284 11285 11286 11287 11288 11289 11290 11291 11292 11293\n",
      " 11294 11295 21632 21633 21634 21635 21636 21637 21638 21639 21640 21764\n",
      " 21765 21766 21767 21768 21769 21986 21987 21988 21989 21990 21991 21992\n",
      " 21993 21994 22028 22029 22030 22031 22032 22033 22193 22194 22195 27888\n",
      " 28247 28248 28249 28250 28251 28252 28270 28361 28362 28363 28364 28365\n",
      " 28366 28367 28368 28369]\n",
      "Europe-Portugal\n",
      "number of sites:\t\t76\n",
      "[[  0.         952.38095238   0.         ...   0.           0.\n",
      "    0.        ]\n",
      " [  0.         952.38095238   0.         ...   0.           0.\n",
      "    0.        ]\n",
      " [  0.         952.38095238   0.         ...   0.           0.\n",
      "    0.        ]\n",
      " ...\n",
      " [  0.         952.38095238   0.         ...   0.           0.\n",
      "   94.15348659]\n",
      " [  0.         952.38095238   0.         ...   0.           0.\n",
      "   62.35962727]\n",
      " [  0.         952.38095238   0.         ...   0.           0.\n",
      "   52.5381959 ]]\n",
      "eliminating sites whose frequency is too low\t0.0648026669950923\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "allele_number: [685, 686, 687, 688, 689, 690, 691, 692, 693, 6512, 6513, 6514, 11282, 11283, 11284, 11285, 11286, 11287, 11288, 11289, 11290, 11291, 11292, 11293, 11294, 11295, 21632, 21633, 21634, 21635, 21636, 21637, 21638, 21639, 21640, 21764, 21765, 21766, 21767, 21768, 21769, 21986, 21987, 21988, 21989, 21990, 21991, 21992, 21993, 21994, 22028, 22029, 22030, 22031, 22032, 22033, 22193, 22194, 22195, 27888, 28247, 28248, 28249, 28250, 28251, 28252, 28270, 28361, 28362, 28363, 28364, 28365, 28366, 28367, 28368, 28369]\n",
      "ref_sites shape: (1, 76)\n",
      "counts shape: (566, 380)\n",
      "times shape: (566,)\n",
      "loading the integrated covariance and the delta x term for location 0, Europe-Portugal---2021-4-20-2022-11-6\t0.027712834009435028\n",
      "[[ 1.97167110e+03 -1.97167110e+03  0.00000000e+00 ...  0.00000000e+00\n",
      "  -2.66639131e-01 -9.60671406e+01]\n",
      " [-1.97167110e+03  1.97167110e+03  0.00000000e+00 ...  0.00000000e+00\n",
      "   2.66639131e-01  9.60671406e+01]\n",
      " [ 0.00000000e+00  0.00000000e+00  0.00000000e+00 ...  0.00000000e+00\n",
      "   0.00000000e+00  0.00000000e+00]\n",
      " ...\n",
      " [ 0.00000000e+00  0.00000000e+00  0.00000000e+00 ...  0.00000000e+00\n",
      "   0.00000000e+00  0.00000000e+00]\n",
      " [-2.66639131e-01  2.66639131e-01  0.00000000e+00 ...  0.00000000e+00\n",
      "   6.06652838e+00 -4.66996114e-01]\n",
      " [-9.60671406e+01  9.60671406e+01  0.00000000e+00 ...  0.00000000e+00\n",
      "  -4.66996114e-01  1.94539450e+03]]\n",
      "[ 0.00000000e+00 -2.27373675e-13  0.00000000e+00  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00 -2.27373675e-13  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00\n",
      "  0.00000000e+00 -2.27373675e-13  0.00000000e+00  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00 -2.27373675e-13\n",
      "  0.00000000e+00  0.00000000e+00 -2.27373675e-13  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00 -2.27373675e-13  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00 -2.27373675e-13  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00 -2.27373675e-13\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00\n",
      " -2.27373675e-13  0.00000000e+00  0.00000000e+00  0.00000000e+00\n",
      " -2.27373675e-13  0.00000000e+00  0.00000000e+00  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00 -2.27373675e-13  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00 -2.27373675e-13\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00 -2.27373675e-13\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00\n",
      "  0.00000000e+00 -2.27373675e-13  0.00000000e+00  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00 -2.27373675e-13  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00 -2.27373675e-13\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00 -2.27373675e-13\n",
      "  0.00000000e+00  3.07144449e+02  0.00000000e+00  0.00000000e+00\n",
      "  0.00000000e+00 -3.07144449e+02  3.07144449e+02  0.00000000e+00\n",
      " -3.07144449e+02  0.00000000e+00  0.00000000e+00  3.07144449e+02\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00 -3.07144449e+02\n",
      "  3.07144449e+02  0.00000000e+00  0.00000000e+00 -3.07144449e+02\n",
      "  0.00000000e+00  3.07144449e+02  0.00000000e+00  0.00000000e+00\n",
      " -3.07144449e+02  0.00000000e+00  3.07144449e+02  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00 -3.07144449e+02  3.07144449e+02\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00 -3.07144449e+02\n",
      "  3.07144449e+02  0.00000000e+00  0.00000000e+00  0.00000000e+00\n",
      " -3.07144449e+02  3.07144449e+02  0.00000000e+00  0.00000000e+00\n",
      "  0.00000000e+00 -3.07144449e+02  9.52380952e+02  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00 -9.52380952e+02  9.52380952e+02\n",
      " -9.52380952e+02  0.00000000e+00  0.00000000e+00  0.00000000e+00\n",
      "  9.52380952e+02  0.00000000e+00 -9.52380952e+02  0.00000000e+00\n",
      "  0.00000000e+00  9.52380952e+02  0.00000000e+00 -9.52380952e+02\n",
      "  0.00000000e+00  0.00000000e+00  9.52380952e+02  0.00000000e+00\n",
      " -9.52380952e+02  0.00000000e+00  0.00000000e+00  9.52380952e+02\n",
      "  0.00000000e+00 -9.52380952e+02  0.00000000e+00  0.00000000e+00\n",
      "  9.52380952e+02  0.00000000e+00 -9.52380952e+02  0.00000000e+00\n",
      "  0.00000000e+00  9.52380952e+02  0.00000000e+00  0.00000000e+00\n",
      "  0.00000000e+00 -9.52380952e+02  9.52380952e+02  0.00000000e+00\n",
      "  0.00000000e+00 -9.52380952e+02  0.00000000e+00 -3.87026728e+02\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00  3.87026728e+02\n",
      " -3.87026728e+02  3.87026728e+02  0.00000000e+00  0.00000000e+00\n",
      "  0.00000000e+00 -3.23217653e+02  0.00000000e+00  3.23217653e+02\n",
      "  0.00000000e+00  0.00000000e+00 -3.23217653e+02  3.23217653e+02\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00 -3.23217653e+02\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00  3.23217653e+02\n",
      " -3.23217653e+02  0.00000000e+00  0.00000000e+00  3.23217653e+02\n",
      "  0.00000000e+00  0.00000000e+00  9.52380952e+02  0.00000000e+00\n",
      " -9.52380952e+02  0.00000000e+00  0.00000000e+00  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00 -2.27373675e-13  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00 -2.27373675e-13  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00\n",
      " -2.27373675e-13 -5.38304376e+02  0.00000000e+00  0.00000000e+00\n",
      "  0.00000000e+00  5.38304376e+02 -5.38304376e+02  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00  5.38304376e+02 -5.38304376e+02\n",
      "  5.38304376e+02  0.00000000e+00  0.00000000e+00  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00\n",
      " -2.27373675e-13  0.00000000e+00  0.00000000e+00  0.00000000e+00\n",
      "  0.00000000e+00 -2.27373675e-13  0.00000000e+00 -2.27373675e-13\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00 -2.27373675e-13  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00\n",
      " -2.27373675e-13  0.00000000e+00  0.00000000e+00  0.00000000e+00\n",
      "  0.00000000e+00 -2.27373675e-13  0.00000000e+00  6.69900566e+02\n",
      " -6.69900566e+02  0.00000000e+00  0.00000000e+00  0.00000000e+00\n",
      " -2.27373675e-13  0.00000000e+00  0.00000000e+00  0.00000000e+00\n",
      "  0.00000000e+00 -2.27373675e-13  0.00000000e+00  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00\n",
      "  0.00000000e+00 -2.27373675e-13  0.00000000e+00  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00 -2.27373675e-13  0.00000000e+00\n",
      "  0.00000000e+00 -2.82480387e+02  0.00000000e+00  2.82480387e+02\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00 -2.27373675e-13\n",
      "  0.00000000e+00  0.00000000e+00 -2.27373675e-13  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00 -2.27373675e-13  0.00000000e+00\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00 -2.27373675e-13\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00\n",
      " -2.27373675e-13  0.00000000e+00  0.00000000e+00  3.95384635e+01\n",
      "  0.00000000e+00 -3.95384635e+01 -6.69507115e+02 -2.82873837e+02\n",
      "  0.00000000e+00  0.00000000e+00  9.52380952e+02  9.52380952e+02\n",
      "  0.00000000e+00  0.00000000e+00 -9.52380952e+02  0.00000000e+00\n",
      "  9.52380952e+02 -9.52380952e+02  0.00000000e+00  0.00000000e+00\n",
      "  0.00000000e+00  9.52380952e+02  0.00000000e+00  0.00000000e+00\n",
      " -9.52380952e+02  0.00000000e+00  9.52380952e+02 -9.52380952e+02\n",
      "  0.00000000e+00  0.00000000e+00  0.00000000e+00  9.52380952e+02\n",
      " -9.52380952e+02  0.00000000e+00  0.00000000e+00  0.00000000e+00\n",
      "  9.52380952e+02  0.00000000e+00 -9.52380952e+02  0.00000000e+00\n",
      "  0.00000000e+00  9.52380952e+02  0.00000000e+00  0.00000000e+00\n",
      " -9.52380952e+02  0.00000000e+00  9.52380952e+02  0.00000000e+00\n",
      " -9.52380952e+02  0.00000000e+00  0.00000000e+00  8.99842756e+02\n",
      " -9.52380952e+02  0.00000000e+00  0.00000000e+00  5.25381959e+01]\n",
      "76\n",
      "regularization applied\n",
      "error bars found\n",
      "selection coefficients found\n",
      "Took 0.008851708000293002 seconds to solve the system of equations\n",
      "max selection coefficient before Gauge transformation:\t0.019412129662094546\n",
      "calculating the inferred coefficients\t0.03903895799885504\n",
      "total time\t0.132048333005514\n"
     ]
    }
   ],
   "source": [
    "%run {inf_script} --data {temp_dir} --timed 1 -o {out_inf_dir} -q 5 --g1 40 --saveCovar --refFile {'ref-index-2023-06-05.csv'}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
